--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -1182,13 +1182,14 @@
     {
         return;
     }
 
     if (!(desk = GetThreadDesktop( GetCurrentThreadId() )) ||
         !GetUserObjectInformationW( desk, UOI_NAME, desk_name, ARRAY_SIZE( desk_name ), NULL ) ||
-        wcscmp( desk_name, L"Default" ))
+        (wcscmp( desk_name, L"Default" ) &&
+         wcscmp( desk_name, L"root" ) /* CW Hack 24270 */))
     {
         return;
     }
 
     SetShellWindow( hwnd );
 }
