--- a/dlls/kernelbase/version.c
+++ b/dlls/kernelbase/version.c
@@ -745,19 +745,42 @@
  */
 DWORD WINAPI GetFileVersionInfoSizeA( LPCSTR filename, LPDWORD handle )
 {
     return GetFileVersionInfoSizeExA( FILE_VER_GET_LOCALISED, filename, handle );
 }
 
+/* CW Hack 24309 */
+static BOOL needs_manor_lords_hack(const WCHAR *dll)
+{
+    static int is_manor_lords = -1;
+    WCHAR path[MAX_PATH], *appname = path, *p;
+
+    if (is_manor_lords == -1)
+    {
+        if (GetModuleFileNameW(NULL, path, MAX_PATH))
+        {
+            if ((p = wcsrchr(path, '/'))) appname = p + 1;
+            if ((p = wcsrchr(path, '\\'))) appname = p + 1;
+            is_manor_lords = !wcsicmp(appname, L"ManorLords.exe");
+        }
+        else
+            is_manor_lords = 0;
+    }
+
+    return is_manor_lords &&
+           (!wcscmp(dll, L"msvcp140_2.dll") || !wcscmp(dll, L"vcruntime140_1.dll"));
+}
+
 /******************************************************************************
  *           GetFileVersionInfoSizeExW       (kernelbase.@)
  */
 DWORD WINAPI GetFileVersionInfoSizeExW( DWORD flags, LPCWSTR filename, LPDWORD ret_handle )
 {
     DWORD len, offset, magic = 1;
     HMODULE hModule;
+    BOOL manor_lords_hack = needs_manor_lords_hack(filename);  /* CW Hack 24309 */
 
     TRACE("(0x%lx,%s,%p)\n", flags, debugstr_w(filename), ret_handle );
 
     if (ret_handle) *ret_handle = 0;
 
     if (!filename)
@@ -791,14 +814,25 @@
             len = SizeofResource( hModule, hRsrc );
         }
         FreeLibrary( hModule );
     }
     else
     {
-        HANDLE handle = CreateFileW( filename, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-                                     NULL, OPEN_EXISTING, 0, 0 );
+        HANDLE handle;
+
+        /* CW Hack 24309 */
+        WCHAR hacked_filename[MAX_PATH];
+        if (manor_lords_hack)
+        {
+            wcscpy(hacked_filename, L"c:\\windows\\system32\\");
+            wcscat(hacked_filename, filename);
+            filename = hacked_filename;
+        }
+
+        handle = CreateFileW( filename, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
+                              NULL, OPEN_EXISTING, 0, 0 );
         if (handle == INVALID_HANDLE_VALUE) return 0;
         magic = find_version_resource( handle, &len, &offset, flags );
         CloseHandle( handle );
     }
 
     switch (magic)
@@ -863,25 +897,26 @@
 BOOL WINAPI GetFileVersionInfoExW( DWORD flags, LPCWSTR filename, DWORD ignored, DWORD datasize, LPVOID data )
 {
     static const char signature[4] = "FE2X";
     DWORD len, offset, magic = 1;
     HMODULE hModule;
     VS_VERSION_INFO_STRUCT32* vvis = data;
+    BOOL manor_lords_hack = needs_manor_lords_hack(filename);  /* CW Hack 24309 */
 
     TRACE("(0x%lx,%s,%ld,size=%ld,data=%p)\n",
           flags, debugstr_w(filename), ignored, datasize, data );
 
     if (!data)
     {
         SetLastError(ERROR_INVALID_DATA);
         return FALSE;
     }
     if (flags & ~FILE_VER_GET_LOCALISED)
         FIXME("flags 0x%lx ignored\n", flags & ~FILE_VER_GET_LOCALISED);
 
-    if ((hModule = LoadLibraryExW( filename, 0, LOAD_LIBRARY_AS_IMAGE_RESOURCE )))
+    if (!manor_lords_hack && (hModule = LoadLibraryExW( filename, 0, LOAD_LIBRARY_AS_IMAGE_RESOURCE )))
     {
         HRSRC hRsrc = NULL;
         if (!(flags & FILE_VER_GET_LOCALISED))
         {
             LANGID english = MAKELANGID( LANG_ENGLISH, SUBLANG_DEFAULT );
             hRsrc = FindResourceExW( hModule, (LPWSTR)VS_FILE_INFO,
@@ -899,14 +934,25 @@
             FreeResource( hMem );
         }
         FreeLibrary( hModule );
     }
     else
     {
-        HANDLE handle = CreateFileW( filename, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-                                     NULL, OPEN_EXISTING, 0, 0 );
+        HANDLE handle;
+
+        /* CW Hack 24309 */
+        WCHAR hacked_filename[MAX_PATH];
+        if (manor_lords_hack)
+        {
+            wcscpy(hacked_filename, L"c:\\windows\\system32\\");
+            wcscat(hacked_filename, filename);
+            filename = hacked_filename;
+        }
+
+        handle = CreateFileW( filename, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
+                              NULL, OPEN_EXISTING, 0, 0 );
         if (handle == INVALID_HANDLE_VALUE) return 0;
         if ((magic = find_version_resource( handle, &len, &offset, flags )))
             len = read_data( handle, offset, data, min( len, datasize ));
         CloseHandle( handle );
     }
 
