--- a/dlls/kernelbase/process.c
+++ b/dlls/kernelbase/process.c
@@ -1712,12 +1751,56 @@
     else ret = SetEnvironmentVariableW( us_name.Buffer, NULL );
     RtlFreeUnicodeString( &us_name );
     return ret;
 }
 
 
+/* CW HACK 19252: Create the vk_swiftshader_icd.json file for Ubisoft Connect */
+static BOOL is_ubisoft(void)
+{
+    WCHAR name[MAX_PATH], *module_exe;
+    if (GetModuleFileNameW( NULL, name, ARRAYSIZE(name) ))
+    {
+        module_exe = wcsrchr( name, '\\' );
+        module_exe = module_exe ? module_exe + 1 : name;
+        if (!wcsicmp( module_exe, L"upc.exe" ) ||
+            !wcsicmp( module_exe, L"UplayWebCore.exe" ))
+            return TRUE;
+    }
+    return FALSE;
+}
+
+static void write_ubisoft_vulkan_json( LPCWSTR value )
+{
+    static const char vk_swiftshader_icd_json[] =
+        "{\"file_format_version\": \"1.0.0\", \"ICD\": {\"library_path\": \".\\\\vk_swiftshader.dll\", \"api_version\": \"1.0.5\"}}";
+    HANDLE h;
+
+    /* VK_ICD_FILENAMES is a semicolon-separated list, but we're expecting
+     * just one path. Bail out if there's more than one.
+     */
+    if (wcschr( value, ';' ))
+        return;
+
+    /* Only write vk_swiftshader_icd.json */
+    if (!wcsstr( value, L"vk_swiftshader_icd.json" ))
+        return;
+
+    /* Create and write the file if it doesn't exist.
+     * This is potentially racy, but upc.exe does this call before launching any helpers.
+     */
+    h = CreateFileW( value, GENERIC_WRITE, 0, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL );
+    if (!h)
+        return;
+
+    WriteFile( h, vk_swiftshader_icd_json, sizeof(vk_swiftshader_icd_json) - 1, NULL, NULL );
+    CloseHandle( h );
+    TRACE( "Created %s for Ubisoft Connect\n", debugstr_w(value) );
+}
+
+
 /***********************************************************************
  *           SetEnvironmentVariableW   (kernelbase.@)
  */
 BOOL WINAPI DECLSPEC_HOTPATCH SetEnvironmentVariableW( LPCWSTR name, LPCWSTR value )
 {
     UNICODE_STRING us_name, us_value;
@@ -1728,12 +1811,18 @@
     if (!name)
     {
         SetLastError( ERROR_ENVVAR_NOT_FOUND );
         return FALSE;
     }
 
+    /* CW HACK 19252: Create the vk_swiftshader_icd.json file if Ubisoft Connect tries
+     * to use it and it doesn't exist.
+     */
+    if (!wcscmp( name, L"VK_ICD_FILENAMES" ) && value && is_ubisoft())
+        write_ubisoft_vulkan_json( value );
+
     RtlInitUnicodeString( &us_name, name );
     if (value)
     {
         RtlInitUnicodeString( &us_value, value );
         status = RtlSetEnvironmentVariable( NULL, &us_name, &us_value );
     }
