--- a/dlls/shcore/main.c
+++ b/dlls/shcore/main.c
@@ -18,30 +18,38 @@
  */
 
 #include <stdarg.h>
 
 #define COBJMACROS
 
+#include "ntstatus.h"
+#define WIN32_NO_STATUS
 #include "windef.h"
 #include "winbase.h"
 #include "wingdi.h"
 #include "winuser.h"
+#include "winternl.h"
 #include "initguid.h"
 #include "ocidl.h"
 #include "featurestagingapi.h"
 #include "shellscalingapi.h"
 #include "shcore.h"
 #define WINSHLWAPI
 #include "shlwapi.h"
+#include "appmodel.h"
 
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(shcore);
 
 static DWORD shcore_tls;
 static IUnknown *process_ref;
+
+/* CW Hack 22310 */
+extern NTSTATUS WINAPI __wine_get_current_process_explicit_app_user_model_id( WCHAR *buffer, INT size );
+extern NTSTATUS WINAPI __wine_set_current_process_explicit_app_user_model_id( const WCHAR *aumid );
 
 BOOL WINAPI DllMain(HINSTANCE instance, DWORD reason, void *reserved)
 {
     TRACE("%p, %lu, %p.\n", instance, reason, reserved);
 
     switch (reason)
@@ -248,21 +256,53 @@
 
     return hr;
 }
 
 HRESULT WINAPI SetCurrentProcessExplicitAppUserModelID(const WCHAR *appid)
 {
-    FIXME("%s: stub\n", debugstr_w(appid));
-    return S_OK;
+    NTSTATUS status;
+    INT len;
+
+    FIXME("%s: semi-stub\n", debugstr_w(appid));
+
+    /* CW Hack 22310 */
+
+    if (!appid) return E_INVALIDARG;
+
+    len = lstrlenW(appid) + 1;
+    /* Native does not enforce the minimum length. */
+    if (len > APPLICATION_USER_MODEL_ID_MAX_LENGTH) return E_INVALIDARG;
+
+    status = __wine_set_current_process_explicit_app_user_model_id(appid);
+    return status ? E_INVALIDARG : S_OK;
 }
 
 HRESULT WINAPI GetCurrentProcessExplicitAppUserModelID(const WCHAR **appid)
 {
-    FIXME("%p: stub\n", appid);
-    *appid = NULL;
-    return E_NOTIMPL;
+    WCHAR *buffer;
+    NTSTATUS status;
+
+    FIXME("%p: semi-stub\n", appid);
+
+    /* CW Hack 22310 */
+
+    if (!appid) return E_INVALIDARG;
+
+    buffer = CoTaskMemAlloc(APPLICATION_USER_MODEL_ID_MAX_LENGTH * sizeof(WCHAR));
+    if (!buffer) return E_OUTOFMEMORY;
+
+    status = __wine_get_current_process_explicit_app_user_model_id(buffer, APPLICATION_USER_MODEL_ID_MAX_LENGTH);
+    if (status)
+    {
+        /* Shouldn't happen; there should be enough space for any AUMID. */
+        CoTaskMemFree(buffer);
+        return E_OUTOFMEMORY;
+    }
+
+    *appid = buffer;
+    return S_OK;
 }
 
 /*************************************************************************
  * CommandLineToArgvW            [SHCORE.@]
  *
  * We must interpret the quotes in the command line to rebuild the argv
--- a/dlls/win32u/driver.c
+++ b/dlls/win32u/driver.c
@@ -1302,12 +1322,16 @@
     /* vulkan support */
     loaderdrv_VulkanInit,
     /* opengl support */
     nulldrv_wine_get_wgl_driver,
     /* thread management */
     nulldrv_ThreadDetach,
+    /* application user model ID support */
+    /* CW Hack 22310 */
+    loaderdrv_SetCurrentProcessExplicitAppUserModelID,
+    loaderdrv_GetCurrentProcessExplicitAppUserModelID,
 };
 
 const struct user_driver_funcs *user_driver = &lazy_load_driver;
 
 /******************************************************************************
  *	     __wine_set_user_driver   (win32u.so)
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -7314,14 +7343,26 @@
     case DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE:
     case DISPLAYCONFIG_DEVICE_INFO_GET_SDR_WHITE_LEVEL:
     default:
         FIXME( "Unimplemented packet type %u.\n", packet->type );
         return STATUS_INVALID_PARAMETER;
     }
+}
+
+NTSTATUS WINAPI __wine_get_current_process_explicit_app_user_model_id( WCHAR *buffer, INT size )
+{
+    /* CW Hack 22310 */
+    return user_driver->pGetCurrentProcessExplicitAppUserModelID( buffer, size );
 }
 
+NTSTATUS WINAPI __wine_set_current_process_explicit_app_user_model_id( const WCHAR *aumid )
+{
+    /* CW Hack 22310 */
+    return user_driver->pSetCurrentProcessExplicitAppUserModelID( aumid );
+}
+
 /******************************************************************************
  *           NtGdiDdDDIEnumAdapters2    (win32u.@)
  */
 NTSTATUS WINAPI NtGdiDdDDIEnumAdapters2( D3DKMT_ENUMADAPTERS2 *desc )
 {
     D3DKMT_OPENADAPTERFROMLUID open_adapter_from_luid;
--- a/dlls/win32u/win32u.spec
+++ b/dlls/win32u/win32u.spec
@@ -1543,6 +1543,11 @@
 
 ################################################################
 # Wine internal extensions
 
 @ stdcall -syscall __wine_get_icm_profile(long long ptr ptr)
 @ stdcall -syscall __wine_get_file_outline_text_metric(wstr ptr ptr ptr)
+
+# shutil
+# CW Hack 22310
+@ stdcall -syscall __wine_get_current_process_explicit_app_user_model_id(ptr long)
+@ stdcall -syscall __wine_set_current_process_explicit_app_user_model_id(wstr)
--- a/dlls/winemac.drv/cocoa_app.h
+++ b/dlls/winemac.drv/cocoa_app.h
@@ -122,21 +124,31 @@
     NSTimeInterval lastSetCursorPositionTime;
 
     id<WineClipCursorHandler> clipCursorHandler;
 
     NSImage* applicationIcon;
 
+    /* CW Hack 22310, 24199 */
+    NSString *explicitAppUserModelID;
+    BOOL terminatingDueToExternalRequest;
+
     BOOL beenActive;
 
     NSMutableSet* windowsBeingDragged;
+
     BOOL useDragNotifications;
+    // CrossOver Hack 10912: Mac Edit menu
+    NSMutableArray* changedKeyEquivalents;
 }
 
 @property (nonatomic) CGEventSourceKeyboardType keyboardType;
 @property (readonly, copy, nonatomic) NSEvent* lastFlagsChanged;
 @property (readonly, nonatomic) BOOL areDisplaysCaptured;
+@property (readonly, nonatomic) BOOL displaysTemporarilyUncapturedForDialog;
+@property (readonly, nonatomic) BOOL temporarilyIgnoreResignEventsForDialog;
+@property (nonatomic, copy) NSString *explicitAppUserModelID;  /* CW Hack 22310 */
 
 @property (readonly) BOOL clippingCursor;
 @property (nonatomic) NSTimeInterval lastSetCursorPositionTime;
 
     + (WineApplicationController*) sharedController;
 
--- a/dlls/winemac.drv/cocoa_app.m
+++ b/dlls/winemac.drv/cocoa_app.m
@@ -39,12 +41,22 @@
 // Internal distributed notification to handle cooperative app activation in Sonoma.
 static NSString* const WineAppWillActivateNotification = @"WineAppWillActivateNotification";
 static NSString* const WineActivatingAppPIDKey = @"ActivatingAppPID";
 static NSString* const WineActivatingAppPrefixKey = @"ActivatingAppPrefix";
 static NSString* const WineActivatingAppConfigDirKey = @"ActivatingAppConfigDir";
 
+/* CW Hack 22310, 24199 */
+// WineExternalQuitRequestNotification is sent on the distributed notification center when an app
+// may need to quit in response to another app quitting via Cocoa. Any app in the same prefix with a
+// corresponding AUMID or EXE name should also quit.
+static NSString* const WineExternalQuitRequestNotification = @"WineExternalQuitRequestNotification";
+static NSString* const WineExternalQuitNotificationAUMIDKey = @"AUMID";
+static NSString* const WineExternalQuitNotificationExeNameKey = @"ExeName";
+static NSString* const WineExternalQuitNotificationSourcePIDKey = @"SourcePID";
+static NSString* const WineExternalQuitNotificationWineConfigDirKey = @"WineConfigDir";
+static NSString* const WineExternalQuitNotificationWinePrefixKey = @"WinePrefix";
 
 int macdrv_err_on;
 
 
 #if !defined(MAC_OS_X_VERSION_10_12) || MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_12
 @interface NSWindow (WineAutoTabbingExtensions)
@@ -115,25 +127,28 @@
 @property (retain, nonatomic) NSImage* applicationIcon;
 @property (readonly, nonatomic) BOOL inputSourceIsInputMethod;
 @property (retain, nonatomic) WineWindow* mouseCaptureWindow;
 
     - (void) setupObservations;
     - (void) applicationDidBecomeActive:(NSNotification *)notification;
+    - (void) handleApplicationShouldTerminateReply:(BOOL)reply;  /* CW Hack 22310 */
 
     static void PerformRequest(void *info);
 
 @end
 
 
 @implementation WineApplicationController
 
     @synthesize keyboardType, lastFlagsChanged;
+    @synthesize displaysTemporarilyUncapturedForDialog, temporarilyIgnoreResignEventsForDialog;
     @synthesize applicationIcon;
     @synthesize cursorFrames, cursorTimer, cursor;
     @synthesize mouseCaptureWindow;
     @synthesize lastSetCursorPositionTime;
+    @synthesize explicitAppUserModelID;  /* CW Hack 22310 */
 
     + (void) initialize
     {
         if (self == [WineApplicationController class])
         {
             NSDictionary<NSString *, id> *defaults =
@@ -1962,13 +2086,67 @@
              suspensionBehavior:NSNotificationSuspensionBehaviorDrop];
 
         [dnc addObserver:self
                 selector:@selector(enabledKeyboardInputSourcesChanged)
                     name:(NSString*)kTISNotifyEnabledKeyboardInputSourcesChanged
                   object:nil];
+
+        /* CW Hack 22310 */
+        [dnc addObserver:self
+                selector:@selector(handleExternalQuitRequest:)
+                    name:WineExternalQuitRequestNotification
+                  object:nil
+      suspensionBehavior:NSNotificationSuspensionBehaviorDeliverImmediately];
+
+        [nc addObserverForName:WineWillShowPermissionDialogNotification
+                        object:nil
+                         queue:[NSOperationQueue mainQueue]
+                    usingBlock:^(NSNotification *note){
+            /* A system-wide permission dialog is about to be displayed which the
+             * user needs to respond to.
+             * If displays are captured for full-screen, they need to be temporarily
+             * uncaptured.
+             * Regardless of display capture, some events also need to be ignored
+             * when the dialog appears, to prevent the app from thinking it's been
+             * switched away from and minimizing itself.
+             */
+            if ([NSApp isActive])
+            {
+                if ([originalDisplayModes count] || displaysCapturedForFullscreen)
+                {
+                    NSNumber* displayID;
+                    for (displayID in originalDisplayModes)
+                    {
+                        CGDisplayModeRef mode = CGDisplayCopyDisplayMode([displayID unsignedIntValue]);
+                        [latentDisplayModes setObject:(id)mode forKey:displayID];
+                        CGDisplayModeRelease(mode);
+                    }
 
+                    CGRestorePermanentDisplayConfiguration();
+                    CGReleaseAllDisplays();
+                    [originalDisplayModes removeAllObjects];
+                    displaysCapturedForFullscreen = FALSE;
+                    displaysTemporarilyUncapturedForDialog = TRUE;
+                }
+                temporarilyIgnoreResignEventsForDialog = TRUE;
+            }
+        }];
+
+        [nc addObserverForName:WineDidShowPermissionDialogNotification
+                        object:nil
+                         queue:[NSOperationQueue mainQueue]
+                    usingBlock:^(NSNotification *note){
+            if (displaysTemporarilyUncapturedForDialog)
+            {
+                [self applicationDidBecomeActive:nil];
+
+                displaysTemporarilyUncapturedForDialog = FALSE;
+            }
+            temporarilyIgnoreResignEventsForDialog = FALSE;
+        }];
+
         if ([NSApplication instancesRespondToSelector:@selector(yieldActivationToApplication:)])
         {
             /* App activation cooperation, starting in macOS 14 Sonoma. */
             [dnc addObserver:self
                     selector:@selector(otherWineAppWillActivate:)
                         name:WineAppWillActivateNotification
@@ -2059,12 +2237,107 @@
               deliverImmediately:YES];
 
         /* This is racy. See the note in otherWineAppWillActivate:. */
         [NSApp activate];
      }
 
+    /* CW Hack 22310, 24199 */
+    - (void) handleExternalQuitRequest:(NSNotification *)notification
+    {
+        NSString *targetAUMID, *targetExeName, *ourExeName;
+        pid_t sourcePID;
+        NSProcessInfo *ourProcess;
+        NSString *ourConfigDir, *otherConfigDir, *ourPrefix, *otherPrefix;
+
+        sourcePID = [notification.userInfo[WineExternalQuitNotificationSourcePIDKey] intValue];
+        ourProcess = [NSProcessInfo processInfo];
+
+        // Ignore requests from ourself
+        if (sourcePID == ourProcess.processIdentifier) return;
+
+        targetAUMID = notification.userInfo[WineExternalQuitNotificationAUMIDKey];
+        if (targetAUMID.length && ![self.explicitAppUserModelID isEqualToString:targetAUMID])
+            return;
+
+        targetExeName = notification.userInfo[WineExternalQuitNotificationExeNameKey];
+        ourExeName = [NSRunningApplication currentApplication].executableURL.lastPathComponent;;
+        if (targetExeName.length && ![ourExeName isEqualToString:targetExeName])
+            return;
+
+        // AUMID or EXE name matches. Make sure it's from the same prefix.
+        ourConfigDir = ourProcess.environment[@"WINECONFIGDIR"];
+        otherConfigDir = notification.userInfo[WineExternalQuitNotificationWineConfigDirKey];
+        if (ourConfigDir.length && otherConfigDir.length &&
+            ![ourConfigDir isEqualToString:otherConfigDir])
+        {
+            return;
+        }
+
+        ourPrefix = ourProcess.environment[@"WINEPREFIX"];
+        otherPrefix = notification.userInfo[WineExternalQuitNotificationWinePrefixKey];
+        if (ourPrefix.length && otherPrefix.length &&
+            ![ourPrefix isEqualToString:otherPrefix])
+        {
+            return;
+        }
+
+        terminatingDueToExternalRequest = YES;
+        [NSApp terminate:NSApp];
+    }
+
+    /* CW Hack 22310, 24199 */
+    - (void) postExternalQuitRequest
+    {
+        static NSArray *whitelistedAUMIDs, *whitelistedExeNames;
+        static dispatch_once_t onceToken;
+        NSDictionary *userInfo;
+        NSProcessInfo *process;
+        NSString *wineConfigDir, *winePrefix, *ourExeName, *targetExeName = @"";
+
+        /* temporarily only enabling this in certain apps that really need it */
+        dispatch_once(&onceToken, ^{
+            whitelistedAUMIDs = [@[
+                @"Valve.Steam.Client",                 /* CW Hack 22310 */
+                @"RockstarGames.SocialClub.UI.Final",  /* CW Hack 23655 */
+            ] retain];
+        });
+
+        ourExeName = [NSRunningApplication currentApplication].executableURL.lastPathComponent;
+        if ([ourExeName isEqualToString:@"EpicGamesLauncher.exe"])
+        {
+            /* CW Hack 24199: Quitting the Epic launcher should also quit the
+               web helpers. */
+            targetExeName = @"EpicWebHelper.exe";
+        }
+        else
+        {
+            if (!self.explicitAppUserModelID.length) return;
+            if (![whitelistedAUMIDs containsObject:self.explicitAppUserModelID]) return;
+        }
+
+        process = [NSProcessInfo processInfo];
+        wineConfigDir = process.environment[@"WINECONFIGDIR"];
+        if (!wineConfigDir) wineConfigDir = @"";
+        winePrefix = process.environment[@"WINEPREFIX"];
+        if (!winePrefix) winePrefix = @"";
+
+        userInfo = @{
+            WineExternalQuitNotificationAUMIDKey: self.explicitAppUserModelID ? self.explicitAppUserModelID : @"",
+            WineExternalQuitNotificationExeNameKey: targetExeName,
+            WineExternalQuitNotificationSourcePIDKey: @([NSProcessInfo processInfo].processIdentifier),
+            WineExternalQuitNotificationWineConfigDirKey: wineConfigDir,
+            WineExternalQuitNotificationWinePrefixKey: winePrefix
+        };
+
+        [[NSDistributedNotificationCenter defaultCenter]
+            postNotificationName:WineExternalQuitRequestNotification
+                          object:nil
+                        userInfo:userInfo
+              deliverImmediately:YES];
+    }
+
     static BOOL InputSourceShouldBeIgnored(TISInputSourceRef inputSource)
     {
         /* Certain system utilities are technically input sources, but we
            shouldn't consider them as such for our purposes. */
         static CFStringRef ignoredIDs[] = {
             /* The "Emoji & Symbols" palette. */
@@ -2310,12 +2586,28 @@
 
         macdrv_release_event(event);
 
         return ret;
     }
 
+    /* CW Hack 22310, 24199 */
+    - (void)handleApplicationShouldTerminateReply:(BOOL)reply
+    {
+        if (reply && !terminatingDueToExternalRequest)
+        {
+            // Normal Cocoa-initiated quit, so potentially tell other related
+            // apps to quit as well.
+            [self postExternalQuitRequest];
+        }
+
+        if (!reply)
+            terminatingDueToExternalRequest = NO;
+
+        [NSApp replyToApplicationShouldTerminate:reply];
+    }
+
     - (void)applicationWillBecomeActive:(NSNotification *)notification
     {
         macdrv_event* event = macdrv_create_event(APP_ACTIVATED, nil);
         event->deliver = 1;
 
         [eventQueuesLock lock];
@@ -2620,28 +2912,39 @@
  *
  * Set the application icon.  The images array contains CGImages.  If
  * there are more than one, then they represent different sizes or
  * color depths from the icon resource.  If images is NULL or empty,
  * restores the default application image.
  */
-void macdrv_set_application_icon(CFArrayRef images)
+void macdrv_set_application_icon(CFArrayRef images, CFURLRef urlRef)
 {
     NSArray* imageArray = (NSArray*)images;
+    NSURL* url = (NSURL*)urlRef;
 
     OnMainThreadAsync(^{
-        [[WineApplicationController sharedController] setApplicationIconFromCGImageArray:imageArray];
+        // CrossOver Hack 13440: Get the icon from the passed-in URL if no images
+        WineApplicationController* controller = [WineApplicationController sharedController];
+        NSImage* image = nil;
+        if (!imageArray && url)
+            image = [[[NSImage alloc] initWithContentsOfURL:url] autorelease];
+        if (imageArray || ![image isValid])
+            [controller setApplicationIconFromCGImageArray:imageArray];
+        else
+            controller.applicationIcon = image;
     });
 }
 
 /***********************************************************************
  *              macdrv_quit_reply
  */
 void macdrv_quit_reply(int reply)
 {
     OnMainThread(^{
-        [NSApp replyToApplicationShouldTerminate:reply];
+        /* CW Hack 22310 - route this through the app controller for
+           app user model ID handling. */
+        [[WineApplicationController sharedController] handleApplicationShouldTerminateReply:reply];
     });
 }
 
 /***********************************************************************
  *              macdrv_using_input_method
  */
@@ -2744,6 +3047,54 @@
     OnMainThread(^{
         ret = [[WineApplicationController sharedController] isAnyWineWindowVisible];
     });
 
     return ret;
 }
+
+/* CW Hack 22310 */
+int macdrv_set_current_process_explicit_app_user_model_id(const UniChar *aumid, size_t length)
+{
+    NSString *str_aumid;
+
+    if (!aumid) return FALSE;
+
+    str_aumid = [NSString stringWithCharacters:aumid length:length];
+    if (!str_aumid) return FALSE;
+
+    OnMainThread(^{
+        [WineApplicationController sharedController].explicitAppUserModelID = str_aumid;
+    });
+
+    return TRUE;
+}
+
+/* CW Hack 22310 */
+int macdrv_get_current_process_explicit_app_user_model_id(UniChar *buffer, size_t size)
+{
+    __block NSString *aumid;
+
+    if (!buffer) return FALSE;
+
+    OnMainThread(^{
+        aumid = [[WineApplicationController sharedController].explicitAppUserModelID copy];
+    });
+
+    if (!aumid || aumid.length == 0)
+    {
+        /* Return empty string if there's no AUMID. */
+        if (size > 0) buffer[0] = '\0';
+        [aumid release];
+        return TRUE;
+    }
+
+    if (aumid.length + 1 > size)
+    {
+        [aumid release];
+        return FALSE;
+    }
+
+    [aumid getCharacters:buffer range:NSMakeRange(0, aumid.length)];
+    buffer[aumid.length] = '\0';
+    [aumid release];
+    return TRUE;
+}
--- a/dlls/winemac.drv/gdi.c
+++ b/dlls/winemac.drv/gdi.c
@@ -272,19 +272,21 @@
     .pNotifyIcon = macdrv_NotifyIcon,
     .pCleanupIcons = macdrv_CleanupIcons,
     .pClipboardWindowProc = macdrv_ClipboardWindowProc,
     .pDesktopWindowProc = macdrv_DesktopWindowProc,
     .pDestroyCursorIcon = macdrv_DestroyCursorIcon,
     .pDestroyWindow = macdrv_DestroyWindow,
+    .pGetCurrentProcessExplicitAppUserModelID = macdrv_GetCurrentProcessExplicitAppUserModelID,  /* CW Hack 22310 */
     .pUpdateDisplayDevices = macdrv_UpdateDisplayDevices,
     .pGetCursorPos = macdrv_GetCursorPos,
     .pGetKeyboardLayoutList = macdrv_GetKeyboardLayoutList,
     .pGetKeyNameText = macdrv_GetKeyNameText,
     .pMapVirtualKeyEx = macdrv_MapVirtualKeyEx,
     .pProcessEvents = macdrv_ProcessEvents,
     .pRegisterHotKey = macdrv_RegisterHotKey,
+    .pSetCurrentProcessExplicitAppUserModelID = macdrv_SetCurrentProcessExplicitAppUserModelID,  /* CW Hack 22310 */
     .pSetCapture = macdrv_SetCapture,
     .pSetCursor = macdrv_SetCursor,
     .pSetCursorPos = macdrv_SetCursorPos,
     .pSetDesktopWindow = macdrv_SetDesktopWindow,
     .pSetFocus = macdrv_SetFocus,
     .pSetLayeredWindowAttributes = macdrv_SetLayeredWindowAttributes,
--- a/dlls/winemac.drv/macdrv.h
+++ b/dlls/winemac.drv/macdrv.h
@@ -132,12 +138,14 @@
 extern BOOL macdrv_SetDeviceGammaRamp(PHYSDEV dev, LPVOID ramp);
 extern BOOL macdrv_ClipCursor(const RECT *clip, BOOL reset);
 extern LRESULT macdrv_NotifyIcon(HWND hwnd, UINT msg, NOTIFYICONDATAW *data);
 extern void macdrv_CleanupIcons(HWND hwnd);
 extern LRESULT macdrv_DesktopWindowProc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam);
 extern void macdrv_DestroyWindow(HWND hwnd);
+extern NTSTATUS macdrv_SetCurrentProcessExplicitAppUserModelID(LPCWSTR aumid);  /* CW Hack 22310 */
+extern NTSTATUS macdrv_GetCurrentProcessExplicitAppUserModelID(LPWSTR buffer, INT size);  /* CW Hack 22310 */
 extern void macdrv_SetDesktopWindow(HWND hwnd);
 extern void macdrv_SetFocus(HWND hwnd);
 extern void macdrv_SetLayeredWindowAttributes(HWND hwnd, COLORREF key, BYTE alpha,
                                               DWORD flags);
 extern void macdrv_SetParent(HWND hwnd, HWND parent, HWND old_parent);
 extern void macdrv_SetWindowRgn(HWND hwnd, HRGN hrgn, BOOL redraw);
--- a/dlls/winemac.drv/macdrv_main.c
+++ b/dlls/winemac.drv/macdrv_main.c
@@ -595,13 +622,32 @@
         }
         break;
     }
     return FALSE;
 }
 
+/* CW Hack 22310 */
+NTSTATUS macdrv_SetCurrentProcessExplicitAppUserModelID(const WCHAR *aumid)
+{
+    if (!macdrv_set_current_process_explicit_app_user_model_id(aumid, lstrlenW(aumid)))
+        return STATUS_INVALID_PARAMETER;
 
+    return 0;
+}
+
+/* CW Hack 22310 */
+NTSTATUS macdrv_GetCurrentProcessExplicitAppUserModelID(WCHAR *buffer, INT size)
+{
+    if (!buffer) return STATUS_INVALID_PARAMETER;
+
+    if (!macdrv_get_current_process_explicit_app_user_model_id(buffer, size))
+        return STATUS_BUFFER_TOO_SMALL;
+
+    return 0;
+}
+
 static NTSTATUS macdrv_quit_result(void *arg)
 {
     struct quit_result_params *params = arg;
     macdrv_quit_reply(params->result);
     return 0;
 }
--- a/dlls/wow64win/user.c
+++ b/dlls/wow64win/user.c
@@ -4901,6 +4901,25 @@
 NTSTATUS WINAPI wow64_NtUserDisplayConfigGetDeviceInfo( UINT *args )
 {
     DISPLAYCONFIG_DEVICE_INFO_HEADER *packet = get_ptr( &args );
 
     return NtUserDisplayConfigGetDeviceInfo( packet );
 }
+
+NTSTATUS WINAPI wow64___wine_get_current_process_explicit_app_user_model_id( UINT *args )
+{
+    /* CW Hack 22310 */
+
+    WCHAR *buffer = get_ptr( &args );
+    INT size = get_ulong( &args );
+
+    return __wine_get_current_process_explicit_app_user_model_id( buffer, size );
+}
+
+NTSTATUS WINAPI wow64___wine_set_current_process_explicit_app_user_model_id( UINT *args )
+{
+    /* CW Hack 22310 */
+
+    const WCHAR *aumid = get_ptr( &args );
+
+    return __wine_set_current_process_explicit_app_user_model_id( aumid );
+}
--- a/include/ntuser.h
+++ b/include/ntuser.h
@@ -1706,8 +1716,12 @@
 
 static inline BOOL NtUserSetRawWindowPos( HWND hwnd, RECT rect, UINT flags, BOOL internal )
 {
     struct set_raw_window_pos_params params = {.rect = rect, .flags = flags, .internal = internal};
     return NtUserCallHwndParam( hwnd, (UINT_PTR)&params, NtUserCallHwndParam_SetRawWindowPos );
 }
+
+/* CW Hack 22310 */
+extern NTSTATUS WINAPI __wine_get_current_process_explicit_app_user_model_id( WCHAR *buffer, INT size );
+extern NTSTATUS WINAPI __wine_set_current_process_explicit_app_user_model_id( const WCHAR *aumid );
 
 #endif /* _NTUSER_ */
--- a/include/wine/gdi_driver.h
+++ b/include/wine/gdi_driver.h
@@ -394,12 +394,16 @@
     /* vulkan support */
     UINT    (*pVulkanInit)(UINT,void *,const struct vulkan_driver_funcs **);
     /* opengl support */
     struct opengl_funcs * (*pwine_get_wgl_driver)(UINT);
     /* thread management */
     void    (*pThreadDetach)(void);
+    /* application user model ID support */
+    /* CW Hack 22310 */
+    NTSTATUS (*pSetCurrentProcessExplicitAppUserModelID)(LPCWSTR);
+    NTSTATUS (*pGetCurrentProcessExplicitAppUserModelID)(LPWSTR,INT);
 };
 
 W32KAPI void __wine_set_user_driver( const struct user_driver_funcs *funcs, UINT version );
 
 W32KAPI BOOL win32u_set_window_pixel_format( HWND hwnd, int format, BOOL internal );
 W32KAPI int win32u_get_window_pixel_format( HWND hwnd );
