--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -3313,12 +3907,23 @@
  *              __wine_ctrl_routine
  */
 NTSTATUS WINAPI __wine_ctrl_routine( void *arg )
 {
     DWORD ret = pCtrlRoutine ? pCtrlRoutine( arg ) : 0;
     RtlExitUserThread( ret );
+}
+
+
+/***********************************************************************
+ *              __wine_unix_call
+ *
+ * CW HACK 22435: Needed by D3DMetal PE DLLs
+ */
+NTSTATUS WINAPI __wine_unix_call_exported( unixlib_handle_t handle, unsigned int code, void *args )
+{
+    return __wine_unix_call( handle, code, args );
 }
 
 
 /***********************************************************************
  *           __wine_unix_spawnvp
  */
