--- a/programs/services/services.c
+++ b/programs/services/services.c
@@ -38,13 +38,17 @@
 #define MAX_SERVICE_NAME 260
 
 WINE_DEFAULT_DEBUG_CHANNEL(service);
 
 struct scmdatabase *active_database;
 
-DWORD service_pipe_timeout = 10000;
+/* CW HACK 20218:
+ * Quadruple service_pipe_timeout, the Rockstar Launcher service takes 20+
+ * seconds to launch under Rosetta.
+ */
+DWORD service_pipe_timeout = 40000;
 DWORD service_kill_timeout = 60000;
 static DWORD default_preshutdown_timeout = 180000;
 static DWORD autostart_delay = 120000;
 static void *environment = NULL;
 static HKEY service_current_key = NULL;
 static HANDLE job_object, job_completion_port;
