--- a/dlls/imm32/ime.c
+++ b/dlls/imm32/ime.c
@@ -202,13 +202,17 @@
     MONITORINFO mon_info;
     INPUTCONTEXT *ctx;
     POINT offset;
     WCHAR *str;
     UINT len;
 
-    if (!(ctx = ImmLockIMC( himc ))) return;
+    if (!(ctx = ImmLockIMC( himc )))
+    {
+        ValidateRect( hwnd, NULL );
+        return;
+    }
 
     hdc = BeginPaint( hwnd, &ps );
 
     GetClientRect( hwnd, &rect );
     FillRect( hdc, &rect, (HBRUSH)(COLOR_WINDOW + 1) );
     new_rect = rect;
--- a/dlls/imm32/imm.c
+++ b/dlls/imm32/imm.c
@@ -3337,7 +3337,27 @@
  */
 BOOL WINAPI CtfImmIsCiceroEnabled(void)
 {
     FIXME("(): stub\n");
     SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
     return FALSE;
+}
+
+/***********************************************************************
+ *      CtfImmHideToolbarWnd (IMM32.@)
+ */
+DWORD WINAPI CtfImmHideToolbarWnd(void)
+{
+    FIXME("(): stub\n");
+    SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
+    return 0;
+}
+
+/***********************************************************************
+ *      CtfImmRestoreToolbarWnd (IMM32.@)
+ */
+DWORD WINAPI CtfImmRestoreToolbarWnd(DWORD unknown)
+{
+    FIXME("%lx: stub\n", unknown);
+    SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
+    return 0;
 }
--- a/dlls/imm32/imm32.spec
+++ b/dlls/imm32/imm32.spec
@@ -4,21 +4,21 @@
 @ stub CtfImmEnterCoInitCountSkipMode
 @ stub CtfImmGenerateMessage
 @ stub CtfImmGetCompatibleKeyboardLayout
 @ stub CtfImmGetGuidAtom
 @ stub CtfImmGetIMEFileName
 @ stub CtfImmGetTMAEFlags
-@ stub CtfImmHideToolbarWnd
+@ stdcall CtfImmHideToolbarWnd()
 @ stdcall CtfImmIsCiceroEnabled()
 @ stub CtfImmIsCiceroStartedInThread
 @ stub CtfImmIsGuidMapEnable
 @ stub CtfImmIsTextFrameServiceDisabled
 @ stub CtfImmLastEnabledWndDestroy
 @ stub CtfImmLeaveCoInitCountSkipMode
 @ stub CtfImmNotify
-@ stub CtfImmRestoreToolbarWnd
+@ stdcall CtfImmRestoreToolbarWnd(long)
 @ stub CtfImmSetAppCompatFlags
 @ stub CtfImmSetCiceroStartInThread
 @ stub CtfImmSetDefaultRemoteKeyboardLayout
 @ stub CtfImmTIMActivate
 @ stub GetKeyboardLayoutCP
 @ stdcall ImmActivateLayout(long)
