--- a/dlls/rpcrt4/rpc_server.c
+++ b/dlls/rpcrt4/rpc_server.c
@@ -698,16 +698,12 @@
     LeaveCriticalSection(&cps->cs);
     rpcrt4_conn_release_and_wait(conn);
     EnterCriticalSection(&cps->cs);
   }
   LeaveCriticalSection(&cps->cs);
 
-  EnterCriticalSection(&listen_cs);
-  CloseHandle(cps->server_thread);
-  cps->server_thread = NULL;
-  LeaveCriticalSection(&listen_cs);
   TRACE("done\n");
   return 0;
 }
 
 /* tells the server thread that the state has changed and waits for it to
  * make the changes */
@@ -1567,21 +1563,25 @@
 
       wait_thread = NULL;
       EnterCriticalSection(&server_cs);
       LIST_FOR_EACH_ENTRY(protseq, &protseqs, RpcServerProtseq, entry)
       {
           if ((wait_thread = protseq->server_thread))
+          {
+              protseq->server_thread = NULL;
               break;
+          }
       }
       LeaveCriticalSection(&server_cs);
       if (!wait_thread)
           break;
 
       TRACE("waiting for thread %lu\n", GetThreadId(wait_thread));
       LeaveCriticalSection(&listen_cs);
       WaitForSingleObject(wait_thread, INFINITE);
+      CloseHandle(wait_thread);
       EnterCriticalSection(&listen_cs);
   }
   if (listen_done_event == event)
   {
       listen_done_event = NULL;
       CloseHandle( event );
--- a/dlls/rpcrt4/rpc_transport.c
+++ b/dlls/rpcrt4/rpc_transport.c
@@ -369,12 +369,25 @@
     return RPC_S_OUT_OF_RESOURCES;
   }
 
   return status;
 }
 
+/*
+ * CXHACK 14391:
+ * ClickToRun hooks NtReadFile and many other APIs. Some of hooked functions
+ * do RPC calls. When we call NtReadFile in such RPC call, it locks a critical
+ * section inside its NtReadFile hook, which causes dead locks.
+ * It most likely works on Windows because RPC calls use LPC ports, so they
+ * don't call NtReadFile.
+ */
+NTSTATUS WINAPI __wine_rpc_NtReadFile(HANDLE hFile, HANDLE hEvent,
+                           PIO_APC_ROUTINE apc, void* apc_user,
+                           PIO_STATUS_BLOCK io_status, void* buffer, ULONG length,
+                                PLARGE_INTEGER offset, PULONG key);
+
 static int rpcrt4_conn_np_read(RpcConnection *conn, void *buffer, unsigned int count)
 {
     RpcConnection_np *connection = (RpcConnection_np *) conn;
     HANDLE event;
     NTSTATUS status;
 
@@ -382,13 +395,13 @@
     if (!event)
         return -1;
 
     if (connection->read_closed)
         status = STATUS_CANCELLED;
     else
-        status = NtReadFile(connection->pipe, event, NULL, NULL, &connection->io_status, buffer, count, NULL, NULL);
+        status = __wine_rpc_NtReadFile(connection->pipe, event, NULL, NULL, &connection->io_status, buffer, count, NULL, NULL);
     if (status == STATUS_PENDING)
     {
         /* check read_closed again before waiting to avoid a race */
         if (connection->read_closed)
         {
             IO_STATUS_BLOCK io_status;
