--- a/.mailmap
+++ b/.mailmap
@@ -15,27 +15,24 @@
 Carlos Lozano <clozano@andago.com>
 Caron Wills <caron@codeweavers.com>
 Charles Davis <cdavis@codeweavers.com>
 Charles Davis <cdavis5x@gmail.com>
 Christopher Gautier <krys@via.ecp.fr>
 David A. Cuthbert <dacut@ece.cmu.edu>
-David Heidelberg <david@ixit.cz>
+David Heidelberger <david@ixit.cz>
 Dennis Björklund <db@zigo.dhs.org>
 Dennis Björklund <dennisb@cs.chalmers.se>
 Dimitrie O. Paun <dimi@bigfoot.com>
 Dimitrie O. Paun <dimi@lattica.com>
 EA Durbin <eadurbin@freebsd.org>
 Eduardo García <eduardo.garcia@gmx.com>
-Elizabeth Figura <z.figura12@gmail.com>
-Elizabeth Figura <zfigura@codeweavers.com>
 Erich Hoover <erich.e.hoover@gmail.com>
 Erich Hoover <erich.e.hoover@wine-staging.com>
 Erwin Wolff <erwinwolffnl@microformatica.com>
 Esme Povirk <vincent@codeweavers.com>
 Evgeny Putin <evcalipt@etersoft.ru>
-Fan WenJie <fanwj@mail.ustc.edu.cn>
 Fatih Aşıcı <fasici@linux-sevenler.org>
 Fernando Martins <fernando@cmartins.nl>
 François Gouget <fgouget@codeweavers.com>
 François Gouget <fgouget@free.fr>
 François Gouget <fgouget@multimania.com>
 François Gouget <fgouget@psn.net>
@@ -155,12 +152,13 @@
 Vladimir Bespalov <vlad.botanic@gmail.com>
 Wei Xie <xiewei@linuxdeepin.com>
 William Poetra Yoga Hadisoesen <williampoetra@yahoo.com>
 Won Kyu Park <wkpark@kldp.org>
 Yuri Kozlov <yuriy.kozlov@gmail.com>
 Zachary Brown <zac@zacbrown.org>
+Zebediah Figura <zfigura@codeweavers.com>
 Zoltan Nagy <nagyzoli@drotposta.hu>
 unidentified <9700584@cs.stadia.fi>
 unidentified <develop@stueben.com>
 unidentified <f.gockel@t-online.de>
 unidentified <ghutchis@cisco.com>
 unidentified <hamish_nospam@yahoo.com>
