--- a/dlls/mountmgr.sys/mountmgr.c
+++ b/dlls/mountmgr.sys/mountmgr.c
@@ -429,17 +429,16 @@
 
     irp->IoStatus.Information = info;
     irp->IoStatus.Status = status;
     IoCompleteRequest( irp, IO_NO_INCREMENT );
 }
 
-/* NT APC called from Unix side to add/remove devices */
-static void CALLBACK device_op( ULONG_PTR arg1, ULONG_PTR arg2, ULONG_PTR arg3 )
+static void device_op( void )
 {
     struct device_info info;
-    struct dequeue_device_op_params params = { arg1, &info };
+    struct dequeue_device_op_params params = { &info };
 
     if (MOUNTMGR_CALL( dequeue_device_op, &params )) return;
 
     switch (info.op)
     {
     case ADD_DOS_DEVICE:
@@ -596,20 +595,19 @@
     IoCompleteRequest( irp, IO_NO_INCREMENT );
     return status;
 }
 
 static DWORD WINAPI device_op_thread( void *arg )
 {
-    for (;;) SleepEx( INFINITE, TRUE );  /* wait for APCs */
+    for (;;) device_op();
     return 0;
 }
 
 static DWORD WINAPI run_loop_thread( void *arg )
 {
-    struct run_loop_params params = {.op_thread = arg, .op_apc = device_op};
-    return MOUNTMGR_CALL( run_loop, &params );
+    return MOUNTMGR_CALL( run_loop, NULL );
 }
 
 
 /* main entry point for the mount point manager driver */
 NTSTATUS WINAPI DriverEntry( DRIVER_OBJECT *driver, UNICODE_STRING *path )
 {
--- a/dlls/mountmgr.sys/mountmgr.h
+++ b/dlls/mountmgr.sys/mountmgr.h
@@ -78,13 +78,13 @@
     SCSI_REDUCED_CDROM_PERIPHERAL = 0x15,
     SCSI_UNKNOWN_PERIPHERAL = ~0
 };
 
 struct scsi_info
 {
-    enum scsi_device_type type;
+    ULONG                 type;
     SCSI_ADDRESS          addr;
     UINT                  init_id;
     char                  driver[64];
     char                  model[64];
 };
 
--- a/dlls/mountmgr.sys/unixlib.c
+++ b/dlls/mountmgr.sys/unixlib.c
@@ -35,19 +35,21 @@
 #endif
 #ifdef HAVE_SYS_STATVFS_H
 # include <sys/statvfs.h>
 #endif
 #include <unistd.h>
 
+#include <pthread.h>
+
+#include <wine/list.h>
+
 #include "unixlib.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(mountmgr);
 
-static struct run_loop_params run_loop_params;
-
 static NTSTATUS errno_to_status( int err )
 {
     TRACE( "errno = %d\n", err );
     switch (err)
     {
     case EAGAIN:    return STATUS_SHARING_VIOLATION;
@@ -128,71 +130,88 @@
         }
         paths++;
     }
     *names = 0;
 }
 
+static pthread_mutex_t device_op_cs = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t device_op_cv = PTHREAD_COND_INITIALIZER;
+static struct list device_op_list = LIST_INIT( device_op_list );
+struct device_info_list
+{
+    struct device_info device_info;
+    struct list entry;
+};
+
 void queue_device_op( enum device_op op, const char *udi, const char *device,
                       const char *mount_point, enum device_type type, const GUID *guid,
                       const char *serial, const char *label, const struct scsi_info *scsi_info )
 {
-    struct device_info *info;
+    struct device_info_list *info;
     char *str, *end;
 
     info = calloc( 1, sizeof(*info) );
-    str = info->str_buffer;
-    end = info->str_buffer + sizeof(info->str_buffer);
-    info->op = op;
-    info->type = type;
+    str = info->device_info.str_buffer;
+    end = info->device_info.str_buffer + sizeof(info->device_info.str_buffer);
+    info->device_info.op = op;
+    info->device_info.type = type;
 #define ADD_STR(s) if (s && str + strlen(s) + 1 <= end) \
     { \
-        info->s = strcpy( str, s ); \
+        info->device_info.s = strcpy( str, s ); \
         str += strlen(str) + 1; \
     }
     ADD_STR(udi);
     ADD_STR(device);
     ADD_STR(mount_point);
     ADD_STR(serial);
     ADD_STR(label);
 #undef ADD_STR
     if (guid)
     {
-        info->guid_buffer = *guid;
-        info->guid = &info->guid_buffer;
+        info->device_info.guid_buffer = *guid;
+        info->device_info.guid = &info->device_info.guid_buffer;
     }
     if (scsi_info)
     {
-        info->scsi_buffer = *scsi_info;
-        info->scsi_info = &info->scsi_buffer;
+        info->device_info.scsi_buffer = *scsi_info;
+        info->device_info.scsi_info = &info->device_info.scsi_buffer;
     }
-    NtQueueApcThread( run_loop_params.op_thread, run_loop_params.op_apc, (ULONG_PTR)info, 0, 0 );
+
+    pthread_mutex_lock(&device_op_cs);
+    list_add_tail(&device_op_list, &info->entry);
+    pthread_cond_signal(&device_op_cv);
+    pthread_mutex_unlock(&device_op_cs);
 }
 
 static NTSTATUS run_loop( void *args )
 {
-    const struct run_loop_params *params = args;
-
-    run_loop_params = *params;
     run_diskarbitration_loop();
     run_dbus_loop();
     return STATUS_SUCCESS;
 }
 
 static NTSTATUS dequeue_device_op( void *args )
 {
     const struct dequeue_device_op_params *params = args;
-    struct device_info *src = (struct device_info *)params->arg;
     struct device_info *dst = params->info;
+    struct device_info_list *src;
 
+    pthread_mutex_lock(&device_op_cs);
+    while (list_empty(&device_op_list))
+        pthread_cond_wait(&device_op_cv, &device_op_cs);
+    src = LIST_ENTRY(list_head(&device_op_list), struct device_info_list, entry);
+    list_remove(&src->entry);
+    pthread_mutex_unlock(&device_op_cs);
+
     /* copy info to client address space and fix up pointers */
-    *dst = *src;
-    if (dst->udi) dst->udi = (char *)dst + (src->udi - (char *)src);
-    if (dst->device) dst->device = (char *)dst + (src->device - (char *)src);
-    if (dst->mount_point) dst->mount_point = (char *)dst + (src->mount_point - (char *)src);
-    if (dst->serial) dst->serial = (char *)dst + (src->serial - (char *)src);
-    if (dst->label) dst->label = (char *)dst + (src->label - (char *)src);
+    *dst = src->device_info;
+    if (dst->udi) dst->udi = (char *)dst + (src->device_info.udi - (char *)&src->device_info);
+    if (dst->device) dst->device = (char *)dst + (src->device_info.device - (char *)&src->device_info);
+    if (dst->mount_point) dst->mount_point = (char *)dst + (src->device_info.mount_point - (char *)&src->device_info);
+    if (dst->serial) dst->serial = (char *)dst + (src->device_info.serial - (char *)&src->device_info);
+    if (dst->label) dst->label = (char *)dst + (src->device_info.label - (char *)&src->device_info);
     if (dst->guid) dst->guid = &dst->guid_buffer;
     if (dst->scsi_info) dst->scsi_info = &dst->scsi_buffer;
 
     free( src );
     return STATUS_SUCCESS;
 }
@@ -498,12 +517,15 @@
     const char *home;
     char *homelink = NULL;
     NTSTATUS status = STATUS_SUCCESS;
 
     if (link && (!strcmp( link, "$HOME" ) || !strncmp( link, "$HOME/", 6 )) && (home = getenv( "HOME" )))
     {
+#ifdef __ANDROID__
+        home = "/sdcard/Download";
+#endif
         link += 5;
         asprintf( &homelink, "%s%s", home, link );
         link = homelink;
     }
 
     /* ignore nonexistent link targets */
@@ -577,6 +599,371 @@
     write_credential,
     delete_credential,
     enumerate_credentials,
 };
 
 C_ASSERT( ARRAYSIZE(__wine_unix_call_funcs) == unix_funcs_count );
+
+#ifdef _WIN64
+
+typedef ULONG PTR32;
+typedef ULONG HANDLE32;
+
+static NTSTATUS wow64_dequeue_device_op(void *args)
+{
+    struct device_info32
+    {
+        ULONG op;
+        ULONG type;
+        PTR32 udi;
+        PTR32 device;
+        PTR32 mount_point;
+        PTR32 serial;
+        PTR32 label;
+        PTR32 guid;
+        PTR32 scsi_info;
+
+        /* buffer space for pointers */
+        GUID guid_buffer;
+        struct scsi_info scsi_buffer;
+        char str_buffer[1024];
+    };
+
+    struct
+    {
+        PTR32 info;
+    } *params32 = args;
+    struct device_info32 *dst = UlongToPtr(params32->info);
+    struct device_info src;
+    struct dequeue_device_op_params params =
+    {
+        &src,
+    };
+    NTSTATUS status;
+
+    status = dequeue_device_op(&params);
+    if (status) return status;
+
+    dst->op = src.op;
+    dst->type = src.type;
+    if (src.udi)
+        dst->udi = PtrToUlong(dst->str_buffer + (src.udi - src.str_buffer));
+    else
+        dst->udi = 0;
+    if (src.device)
+        dst->device = PtrToUlong(dst->str_buffer + (src.device - src.str_buffer));
+    else
+        dst->device = 0;
+    if (src.mount_point)
+        dst->mount_point = PtrToUlong(dst->str_buffer + (src.mount_point - src.str_buffer));
+    else
+        dst->mount_point = 0;
+    if (src.serial)
+        dst->serial = PtrToUlong(dst->str_buffer + (src.serial - src.str_buffer));
+    else
+        dst->serial = 0;
+    if (src.label)
+        dst->label = PtrToUlong(dst->str_buffer + (src.label - src.str_buffer));
+    else
+        dst->label = 0;
+    if (src.guid)
+        dst->guid = PtrToUlong(&dst->guid_buffer);
+    else
+        dst->guid = 0;
+    if (src.scsi_info)
+        dst->scsi_info = PtrToUlong(&dst->scsi_info);
+    else
+        dst->scsi_info = 0;
+    dst->guid_buffer = src.guid_buffer;
+    dst->scsi_buffer = src.scsi_buffer;
+    memcpy(dst->str_buffer, src.str_buffer, sizeof(dst->str_buffer));
+
+    return STATUS_SUCCESS;
+}
+
+static NTSTATUS wow64_add_drive(void *args)
+{
+    struct
+    {
+        PTR32 device;
+        ULONG type;
+        PTR32 letter;
+    } *params32 = args;
+    struct add_drive_params params =
+    {
+        ULongToPtr(params32->device),
+        params32->type,
+        ULongToPtr(params32->letter),
+    };
+    return add_drive(&params);
+}
+
+static NTSTATUS wow64_get_dosdev_symlink(void *args)
+{
+    struct
+    {
+        PTR32 dev;
+        PTR32 dest;
+        ULONG size;
+    } *params32 = args;
+    struct get_dosdev_symlink_params params =
+    {
+        ULongToPtr(params32->dev),
+        ULongToPtr(params32->dest),
+        params32->size,
+    };
+    return get_dosdev_symlink(&params);
+}
+
+static NTSTATUS wow64_set_dosdev_symlink(void *args)
+{
+    struct
+    {
+        PTR32 dev;
+        PTR32 dest;
+    } *params32 = args;
+    struct set_dosdev_symlink_params params =
+    {
+        ULongToPtr(params32->dev),
+        ULongToPtr(params32->dest),
+    };
+    return set_dosdev_symlink(&params);
+}
+
+static NTSTATUS wow64_get_volume_size_info(void *args)
+{
+    struct
+    {
+        PTR32 unix_mount;
+        PTR32 info;
+    } *params32 = args;
+    struct get_volume_size_info_params params =
+    {
+        ULongToPtr(params32->unix_mount),
+        ULongToPtr(params32->info),
+    };
+    return get_volume_size_info(&params);
+}
+
+static NTSTATUS wow64_get_volume_dos_devices(void *args)
+{
+    struct
+    {
+        PTR32 mount_point;
+        PTR32 dosdev;
+    } *params32 = args;
+    struct get_volume_dos_devices_params params =
+    {
+        ULongToPtr(params32->mount_point),
+        ULongToPtr(params32->dosdev),
+    };
+    return get_volume_dos_devices(&params);
+}
+
+static NTSTATUS wow64_read_volume_file(void *args)
+{
+    struct
+    {
+        PTR32 volume;
+        PTR32 file;
+        PTR32 buffer;
+        PTR32 size;
+    } *params32 = args;
+    struct read_volume_file_params params =
+    {
+        ULongToPtr(params32->volume),
+        ULongToPtr(params32->file),
+        ULongToPtr(params32->buffer),
+        ULongToPtr(params32->size),
+    };
+    return read_volume_file(&params);
+}
+
+static NTSTATUS wow64_match_unixdev(void *args)
+{
+    struct
+    {
+        PTR32 device;
+        ULONGLONG unix_dev;
+    } *params32 = args;
+    struct match_unixdev_params params =
+    {
+        ULongToPtr(params32->device),
+        params32->unix_dev,
+    };
+    return match_unixdev(&params);
+}
+
+static NTSTATUS wow64_check_device_access(void *args)
+{
+    struct
+    {
+        PTR32 unix_device;
+    } *params32 = args;
+    return check_device_access(ULongToPtr(params32->unix_device));
+}
+
+static NTSTATUS wow64_detect_serial_ports(void *args)
+{
+    struct
+    {
+        PTR32 names;
+        ULONG size;
+    } *params32 = args;
+    struct detect_ports_params params =
+    {
+        ULongToPtr(params32->names),
+        params32->size,
+    };
+    return detect_serial_ports(&params);
+}
+
+static NTSTATUS wow64_detect_parallel_ports(void *args)
+{
+    struct
+    {
+        PTR32 names;
+        ULONG size;
+    } *params32 = args;
+    struct detect_ports_params params =
+    {
+        ULongToPtr(params32->names),
+        params32->size,
+    };
+    return detect_parallel_ports(&params);
+}
+
+static NTSTATUS wow64_set_shell_folder(void *args)
+{
+    struct
+    {
+        PTR32 folder;
+        PTR32 backup;
+        PTR32 link;
+    } *params32 = args;
+    struct set_shell_folder_params params =
+    {
+        ULongToPtr(params32->folder),
+        ULongToPtr(params32->backup),
+        ULongToPtr(params32->link),
+    };
+    return set_shell_folder(&params);
+}
+
+static NTSTATUS wow64_get_shell_folder(void *args)
+{
+    struct
+    {
+        PTR32 folder;
+        PTR32 buffer;
+        ULONG size;
+    } *params32 = args;
+    struct get_shell_folder_params params =
+    {
+        ULongToPtr(params32->folder),
+        ULongToPtr(params32->buffer),
+        params32->size,
+    };
+    return get_shell_folder(&params);
+}
+
+static NTSTATUS wow64_dhcp_request(void *args)
+{
+    struct
+    {
+        PTR32 unix_name;
+        PTR32 req;
+        PTR32 buffer;
+        ULONG offset;
+        ULONG size;
+        PTR32 ret_size;
+    } *params32 = args;
+    struct dhcp_request_params params =
+    {
+        ULongToPtr(params32->unix_name),
+        ULongToPtr(params32->req),
+        ULongToPtr(params32->buffer),
+        params32->offset,
+        params32->size,
+        ULongToPtr(params32->ret_size),
+    };
+    return dhcp_request(&params);
+}
+
+static void wow64_get_ioctl_params(void *args, struct ioctl_params *params)
+{
+    struct
+    {
+        PTR32 buff;
+        ULONG insize;
+        ULONG outsize;
+        PTR32 info;
+    } *params32 = args;
+
+    params->buff = ULongToPtr(params32->buff);
+    params->insize = params32->insize;
+    params->outsize = params32->outsize;
+    params->info = ULongToPtr(params32->info);
+}
+
+static NTSTATUS wow64_query_symbol_file(void *args)
+{
+    struct ioctl_params params;
+    wow64_get_ioctl_params(args, &params);
+    return query_symbol_file(&params);
+}
+
+static NTSTATUS wow64_read_credential(void *args)
+{
+    struct ioctl_params params;
+    wow64_get_ioctl_params(args, &params);
+    return read_credential(&params);
+}
+
+static NTSTATUS wow64_write_credential(void *args)
+{
+    struct ioctl_params params;
+    wow64_get_ioctl_params(args, &params);
+    return write_credential(&params);
+}
+
+static NTSTATUS wow64_delete_credential(void *args)
+{
+    struct ioctl_params params;
+    wow64_get_ioctl_params(args, &params);
+    return delete_credential(&params);
+}
+
+static NTSTATUS wow64_enumerate_credentials(void *args)
+{
+    struct ioctl_params params;
+    wow64_get_ioctl_params(args, &params);
+    return enumerate_credentials(&params);
+}
+
+const unixlib_entry_t __wine_unix_call_wow64_funcs[] =
+{
+    run_loop,
+    wow64_dequeue_device_op,
+    wow64_add_drive,
+    wow64_get_dosdev_symlink,
+    wow64_set_dosdev_symlink,
+    wow64_get_volume_size_info,
+    wow64_get_volume_dos_devices,
+    wow64_read_volume_file,
+    wow64_match_unixdev,
+    wow64_check_device_access,
+    wow64_detect_serial_ports,
+    wow64_detect_parallel_ports,
+    wow64_set_shell_folder,
+    wow64_get_shell_folder,
+    wow64_dhcp_request,
+    wow64_query_symbol_file,
+    wow64_read_credential,
+    wow64_write_credential,
+    wow64_delete_credential,
+    wow64_enumerate_credentials,
+};
+
+C_ASSERT( ARRAYSIZE(__wine_unix_call_funcs) == unix_funcs_count );
+
+#endif  /* _WIN64 */
--- a/dlls/mountmgr.sys/unixlib.h
+++ b/dlls/mountmgr.sys/unixlib.h
@@ -43,21 +43,14 @@
     /* buffer space for pointers */
     GUID              guid_buffer;
     struct scsi_info  scsi_buffer;
     char              str_buffer[1024];
 };
 
-struct run_loop_params
-{
-    HANDLE     op_thread;
-    PNTAPCFUNC op_apc;
-};
-
 struct dequeue_device_op_params
 {
-    ULONG_PTR arg;
     struct device_info *info;
 };
 
 struct add_drive_params
 {
     const char *device;
