--- a/dlls/msvcp90/misc.c
+++ b/dlls/msvcp90/misc.c
@@ -712,16 +712,16 @@
 #define MTX_TIMED 0x4
 #define MTX_RECURSIVE 0x100
 #define MTX_LOCKED 3
 typedef struct
 {
     DWORD flags;
-    cs cs;
 #if _MSVCP_VER >= 140
     ULONG_PTR unknown;
 #endif
+    cs cs;
     DWORD thread_id;
     DWORD count;
 } *_Mtx_t;
 
 #if _MSVCP_VER >= 140
 typedef _Mtx_t _Mtx_arg_t;
@@ -736,12 +736,15 @@
 void __cdecl _Mtx_init_in_situ(_Mtx_t mtx, int flags)
 {
     if(flags & ~(MTX_PLAIN | MTX_TRY | MTX_TIMED | MTX_RECURSIVE))
         FIXME("unknown flags ignored: %x\n", flags);
 
     mtx->flags = flags;
+#if _MSVCP_VER >= 140
+    mtx->unknown = 0;
+#endif
     cs_init(&mtx->cs);
     mtx->thread_id = -1;
     mtx->count = 0;
 }
 
 int __cdecl _Mtx_init(_Mtx_t *mtx, int flags)
@@ -826,12 +829,15 @@
 }
 
 #define CND_TIMEDOUT 2
 
 typedef struct
 {
+#if _MSVCP_VER >= 140
+    ULONG_PTR unknown;
+#endif
     cv cv;
 } *_Cnd_t;
 
 #if _MSVCP_VER >= 140
 typedef _Cnd_t _Cnd_arg_t;
 #define CND_T_FROM_ARG(c)   (c)
@@ -841,12 +847,15 @@
 #define CND_T_FROM_ARG(c)   (*(c))
 #define CND_T_TO_ARG(c)     (&(c))
 #endif
 
 void __cdecl _Cnd_init_in_situ(_Cnd_t cnd)
 {
+#if _MSVCP_VER >= 140
+    cnd->unknown = 0;
+#endif
     cv_init(&cnd->cv);
 }
 
 int __cdecl _Cnd_init(_Cnd_t *cnd)
 {
     *cnd = operator_new(sizeof(**cnd));
