--- a/dlls/msvcp140/tests/msvcp140.c
+++ b/dlls/msvcp140/tests/msvcp140.c
@@ -163,22 +163,29 @@
     cs_queue unk_active;
     void *unknown[2];
     cs_queue *head;
     void *tail;
 } critical_section;
 
-typedef struct
-{
+typedef union {
+    critical_section conc;
+    SRWLOCK win;
+} cs;
+
+typedef struct {
     DWORD flags;
-    critical_section cs;
     ULONG_PTR unknown;
+    cs cs;
     DWORD thread_id;
     DWORD count;
 } *_Mtx_t;
 
-typedef void *_Cnd_t;
+typedef struct {
+    ULONG_PTR unknown;
+    CONDITION_VARIABLE cv;
+} *_Cnd_t;
 
 typedef struct {
     __time64_t sec;
     int nsec;
 } xtime;
 
@@ -215,12 +222,14 @@
 static int (__cdecl *p__Reschedule_chore)(const _Threadpool_chore*);
 static void (__cdecl *p__Release_chore)(_Threadpool_chore*);
 static int (__cdecl *p__Mtx_init)(_Mtx_t*, int);
 static void (__cdecl *p__Mtx_destroy)(_Mtx_t);
 static int (__cdecl *p__Mtx_lock)(_Mtx_t);
 static int (__cdecl *p__Mtx_unlock)(_Mtx_t);
+static void (__cdecl *p__Mtx_clear_owner)(_Mtx_t);
+static void (__cdecl *p__Mtx_reset_owner)(_Mtx_t);
 static int (__cdecl *p__Cnd_init)(_Cnd_t*);
 static void (__cdecl *p__Cnd_destroy)(_Cnd_t);
 static int (__cdecl *p__Cnd_wait)(_Cnd_t, _Mtx_t);
 static int (__cdecl *p__Cnd_timedwait)(_Cnd_t, _Mtx_t, const xtime*);
 static int (__cdecl *p__Cnd_broadcast)(_Cnd_t);
 static int (__cdecl *p__Cnd_signal)(_Cnd_t);
@@ -351,12 +360,14 @@
     }
 
     SET(p__Mtx_init, "_Mtx_init");
     SET(p__Mtx_destroy, "_Mtx_destroy");
     SET(p__Mtx_lock, "_Mtx_lock");
     SET(p__Mtx_unlock, "_Mtx_unlock");
+    SET(p__Mtx_clear_owner, "_Mtx_clear_owner");
+    SET(p__Mtx_reset_owner, "_Mtx_reset_owner");
     SET(p__Cnd_init, "_Cnd_init");
     SET(p__Cnd_destroy, "_Cnd_destroy");
     SET(p__Cnd_wait, "_Cnd_wait");
     SET(p__Cnd_timedwait, "_Cnd_timedwait");
     SET(p__Cnd_broadcast, "_Cnd_broadcast");
     SET(p__Cnd_signal, "_Cnd_signal");
@@ -1497,34 +1508,42 @@
     LONG started;
     int thread_no;
 
     _Cnd_t cnd;
     _Mtx_t mtx;
     BOOL timed_wait;
+    BOOL use_cnd_func;
 };
 
 static int __cdecl cnd_wait_thread(void *arg)
 {
     struct cndmtx *cm = arg;
     int r;
 
     p__Mtx_lock(cm->mtx);
 
     if(InterlockedIncrement(&cm->started) == cm->thread_no)
         SetEvent(cm->initialized);
 
-    if(cm->timed_wait) {
-        xtime xt;
+    if(cm->use_cnd_func) {
+        if(cm->timed_wait) {
+            xtime xt;
+            p_xtime_get(&xt, 1);
+            xt.sec += 2;
 
-        p_xtime_get(&xt, 1);
-        xt.sec += 2;
-        r = p__Cnd_timedwait(cm->cnd, cm->mtx, &xt);
-        ok(!r, "timed wait failed\n");
-    } else {
-        r = p__Cnd_wait(cm->cnd, cm->mtx);
+            r = p__Cnd_timedwait(cm->cnd, cm->mtx, &xt);
+        } else {
+            r = p__Cnd_wait(cm->cnd, cm->mtx);
+        }
         ok(!r, "wait failed\n");
+    } else {
+        p__Mtx_clear_owner(cm->mtx);
+        r = SleepConditionVariableSRW(&cm->cnd->cv, &cm->mtx->cs.win,
+                                      cm->timed_wait ? 2000 : INFINITE, 0);
+        ok(r, "wait failed\n");
+        p__Mtx_reset_owner(cm->mtx);
     }
 
     p__Mtx_unlock(cm->mtx);
     return 0;
 }
 
@@ -1549,18 +1568,33 @@
     cm.initialized = CreateEventW(NULL, FALSE, FALSE, NULL);
     cm.started = 0;
     cm.thread_no = 1;
     cm.cnd = cnd;
     cm.mtx = mtx;
     cm.timed_wait = FALSE;
+    cm.use_cnd_func = TRUE;
+    p__Thrd_create(&threads[0], cnd_wait_thread, (void*)&cm);
+
+    WaitForSingleObject(cm.initialized, INFINITE);
+    p__Mtx_lock(mtx);
+    p__Mtx_unlock(mtx);
+
+    /* signal cnd function with kernel function */
+    WakeConditionVariable(&cm.cnd->cv);
+    p__Thrd_join(threads[0], NULL);
+
+    cm.started = 0;
+    cm.thread_no = 1;
+    cm.use_cnd_func = FALSE;
     p__Thrd_create(&threads[0], cnd_wait_thread, (void*)&cm);
 
     WaitForSingleObject(cm.initialized, INFINITE);
     p__Mtx_lock(mtx);
     p__Mtx_unlock(mtx);
 
+    /* signal kernel functions with cnd function */
     r = p__Cnd_signal(cm.cnd);
     ok(!r, "failed to signal\n");
     p__Thrd_join(threads[0], NULL);
 
     /* test _Cnd_timedwait time out */
     p__Mtx_lock(mtx);
@@ -1577,33 +1611,63 @@
     ok(r == 2, "should have timed out\n");
     ok(diff > 1000 - TIMEDELTA, "got %d\n", diff);
 
     /* test _Cnd_timedwait */
     cm.started = 0;
     cm.timed_wait = TRUE;
+    cm.use_cnd_func = TRUE;
     p__Thrd_create(&threads[0], cnd_wait_thread, (void*)&cm);
 
     WaitForSingleObject(cm.initialized, INFINITE);
     p__Mtx_lock(mtx);
     p__Mtx_unlock(mtx);
 
+    /* signal cnd function with kernel function */
+    WakeConditionVariable(&cm.cnd->cv);
+    p__Thrd_join(threads[0], NULL);
+
+    cm.started = 0;
+    cm.use_cnd_func = FALSE;
+    p__Thrd_create(&threads[0], cnd_wait_thread, (void*)&cm);
+
+    WaitForSingleObject(cm.initialized, INFINITE);
+    p__Mtx_lock(mtx);
+    p__Mtx_unlock(mtx);
+
+    /* signal kernel functions with cnd function */
     r = p__Cnd_signal(cm.cnd);
     ok(!r, "failed to signal\n");
     p__Thrd_join(threads[0], NULL);
 
     /* test _Cnd_broadcast */
     cm.started = 0;
+    cm.timed_wait = FALSE;
+    cm.use_cnd_func = TRUE;
     cm.thread_no = NUM_THREADS;
+    for(i = 0; i < cm.thread_no; i++)
+        p__Thrd_create(&threads[i], cnd_wait_thread, (void*)&cm);
 
+    WaitForSingleObject(cm.initialized, INFINITE);
+    p__Mtx_lock(mtx);
+    p__Mtx_unlock(mtx);
+
+    /* signal cnd function with kernel function */
+    WakeAllConditionVariable(&cm.cnd->cv);
     for(i = 0; i < cm.thread_no; i++)
+        p__Thrd_join(threads[i], NULL);
+
+    cm.started = 0;
+    cm.use_cnd_func = FALSE;
+    for(i = 0; i < cm.thread_no; i++)
         p__Thrd_create(&threads[i], cnd_wait_thread, (void*)&cm);
 
     WaitForSingleObject(cm.initialized, INFINITE);
     p__Mtx_lock(mtx);
     p__Mtx_unlock(mtx);
 
+    /* signal kernel functions with cnd function */
     r = p__Cnd_broadcast(cnd);
     ok(!r, "failed to broadcast\n");
     for(i = 0; i < cm.thread_no; i++)
         p__Thrd_join(threads[i], NULL);
 
     p__Cnd_destroy(cnd);
@@ -1679,21 +1743,25 @@
 
     r = p__Mtx_init(&mtx, 0);
     ok(!r, "failed to init mtx\n");
 
     ok(mtx->thread_id == -1, "mtx.thread_id = %lx\n", mtx->thread_id);
     ok(mtx->count == 0, "mtx.count = %lx\n", mtx->count);
+    ok(mtx->cs.win.Ptr == 0, "mtx.cs == %p\n", mtx->cs.win.Ptr);
     p__Mtx_lock(mtx);
     ok(mtx->thread_id == GetCurrentThreadId(), "mtx.thread_id = %lx\n", mtx->thread_id);
     ok(mtx->count == 1, "mtx.count = %lx\n", mtx->count);
+    ok(mtx->cs.win.Ptr != 0, "mtx.cs == %p\n", mtx->cs.win.Ptr);
     p__Mtx_lock(mtx);
     ok(mtx->thread_id == GetCurrentThreadId(), "mtx.thread_id = %lx\n", mtx->thread_id);
     ok(mtx->count == 1, "mtx.count = %lx\n", mtx->count);
+    ok(mtx->cs.win.Ptr != 0, "mtx.cs == %p\n", mtx->cs.win.Ptr);
     p__Mtx_unlock(mtx);
     ok(mtx->thread_id == -1, "mtx.thread_id = %lx\n", mtx->thread_id);
     ok(mtx->count == 0, "mtx.count = %lx\n", mtx->count);
+    ok(mtx->cs.win.Ptr == 0, "mtx.cs == %p\n", mtx->cs.win.Ptr);
     p__Mtx_unlock(mtx);
     ok(mtx->thread_id == -1, "mtx.thread_id = %lx\n", mtx->thread_id);
     ok(mtx->count == -1, "mtx.count = %lx\n", mtx->count);
     p__Mtx_unlock(mtx);
     ok(mtx->thread_id == -1, "mtx.thread_id = %lx\n", mtx->thread_id);
     ok(mtx->count == -2, "mtx.count = %lx\n", mtx->count);
