--- a/dlls/winecoreaudio.drv/Makefile.in
+++ b/dlls/winecoreaudio.drv/Makefile.in
@@ -1,11 +1,12 @@
 MODULE    = winecoreaudio.drv
 UNIXLIB   = winecoreaudio.so
 IMPORTS   = uuid ole32 user32 advapi32
 DELAYIMPORTS = winmm
-UNIX_LIBS    = $(COREAUDIO_LIBS)
+UNIX_LIBS    = $(COREAUDIO_LIBS) -framework AppKit $(AVFOUNDATION_LIBS)
 
 SOURCES = \
+	authorization.m \
 	coreaudio.c \
 	coremidi.c \
 	midi.c \
 	mmdevdrv.c
--- a/dlls/winecoreaudio.drv/coreaudio.c
+++ b/dlls/winecoreaudio.drv/coreaudio.c
@@ -65,12 +65,13 @@
 #include "initguid.h"
 #include "audioclient.h"
 #include "wine/debug.h"
 #include "wine/unixlib.h"
 
 #include "unixlib.h"
+#include "coreaudio_cocoa.h"
 
 #if !defined(MAC_OS_VERSION_12_0) || MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_VERSION_12_0
 #define kAudioObjectPropertyElementMain kAudioObjectPropertyElementMaster
 #endif
 
 #if defined(MAC_OS_X_VERSION_10_12) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_12
@@ -80,12 +81,13 @@
 typedef OSSpinLock                  os_unfair_lock;
 #define os_unfair_lock_lock(lock)   OSSpinLockLock(lock)
 #define os_unfair_lock_unlock(lock) OSSpinLockUnlock(lock)
 #endif
 
 WINE_DEFAULT_DEBUG_CHANNEL(coreaudio);
+WINE_DECLARE_DEBUG_CHANNEL(winediag);
 
 struct coreaudio_stream
 {
     os_unfair_lock lock;
     AudioComponentInstance unit;
     AudioConverterRef converter;
@@ -766,12 +768,28 @@
                                   kAudioUnitScope_Input, 0, &input, sizeof(input));
     }
     if(sc != noErr){
         WARN("Couldn't set callback: %x\n", (int)sc);
         params->result = osstatus_to_hresult(sc);
         goto end;
+    }
+
+    if (stream->flow == eCapture) {
+        /* On 10.14 and later:
+         * Check the audio capture/microphone authorization status, and explicitly
+         * request it if the user hasn't previously granted or denied permission.
+         */
+        int authstatus = CoreAudio_get_capture_authorization_status();
+        TRACE("Audio capture authorization status: %d\n", authstatus);
+
+        if (authstatus == NOT_DETERMINED)
+            authstatus = CoreAudio_request_capture_authorization();
+
+        if (authstatus == 0)
+            ERR_(winediag)("Microphone/audio capture permission was denied. "
+                           "This can be enabled under Security & Privacy in System Preferences.\n");
     }
 
     sc = AudioUnitInitialize(stream->unit);
     if(sc != noErr){
         WARN("Couldn't initialize: %x\n", (int)sc);
         params->result = osstatus_to_hresult(sc);
