--- a/dlls/mshtml/htmlstyle.c
+++ b/dlls/mshtml/htmlstyle.c
@@ -404,12 +404,17 @@
     },
     {
         L"column-width",
         DISPID_IHTMLCSSSTYLEDECLARATION2_COLUMNWIDTH,
         DISPID_UNKNOWN,
         ATTR_COMPAT_IE10
+    },
+    {
+        L"content",
+        DISPID_IHTMLCSSSTYLEDECLARATION_CONTENT,
+        DISPID_IHTMLSTYLE6_CONTENT
     },
     {
         L"cursor",
         DISPID_IHTMLCSSSTYLEDECLARATION_CURSOR,
         DISPID_A_CURSOR,
         ATTR_BUILTIN_NAME
@@ -4179,21 +4184,21 @@
 
 DISPEX_IDISPATCH_IMPL(HTMLStyle6, IHTMLStyle6, impl_from_IHTMLStyle6(iface)->css_style.dispex)
 
 static HRESULT WINAPI HTMLStyle6_put_content(IHTMLStyle6 *iface, BSTR v)
 {
     HTMLStyle *This = impl_from_IHTMLStyle6(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+    return set_style_property(&This->css_style, STYLEID_CONTENT, v);
 }
 
 static HRESULT WINAPI HTMLStyle6_get_content(IHTMLStyle6 *iface, BSTR *p)
 {
     HTMLStyle *This = impl_from_IHTMLStyle6(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, p);
+    return get_style_property(&This->css_style, STYLEID_CONTENT, p);
 }
 
 static HRESULT WINAPI HTMLStyle6_put_contentSide(IHTMLStyle6 *iface, BSTR v)
 {
     HTMLStyle *This = impl_from_IHTMLStyle6(iface);
     FIXME("(%p)->(%s)\n", This, debugstr_w(v));
@@ -6496,21 +6501,21 @@
     return get_style_property_var(This, STYLEID_MAX_WIDTH, p);
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_put_content(IHTMLCSSStyleDeclaration *iface, BSTR v)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+    return set_style_property(This, STYLEID_CONTENT, v);
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_get_content(IHTMLCSSStyleDeclaration *iface, BSTR *p)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, p);
+    return get_style_property(This, STYLEID_CONTENT, p);
 }
 
 static HRESULT WINAPI HTMLCSSStyleDeclaration_put_captionSide(IHTMLCSSStyleDeclaration *iface, BSTR v)
 {
     CSSStyle *This = impl_from_IHTMLCSSStyleDeclaration(iface);
     FIXME("(%p)->(%s)\n", This, debugstr_w(v));
--- a/dlls/mshtml/htmlstyle.h
+++ b/dlls/mshtml/htmlstyle.h
@@ -88,12 +88,13 @@
     STYLEID_COLUMN_RULE,
     STYLEID_COLUMN_RULE_COLOR,
     STYLEID_COLUMN_RULE_STYLE,
     STYLEID_COLUMN_RULE_WIDTH,
     STYLEID_COLUMN_SPAN,
     STYLEID_COLUMN_WIDTH,
+    STYLEID_CONTENT,
     STYLEID_CURSOR,
     STYLEID_DIRECTION,
     STYLEID_DISPLAY,
     STYLEID_FILTER,
     STYLEID_FLOAT,
     STYLEID_FONT_FAMILY,
--- a/dlls/mshtml/mshtml.inf
+++ b/dlls/mshtml/mshtml.inf
@@ -113,13 +113,14 @@
 HKCR,"MIME\Database\Content Type\image/gif","Image Filter CLSID",,"%CLSID_IImageDecodeFilter%"
 HKCR,"MIME\Database\Content Type\image/gif\Bits","0",1,04,00,00,00,FF,FF,FF,FF,47,49,46,38
 HKCR,".gif",,2,"giffile"
 HKCR,".gif","Content Type",,"image/gif"
 HKCR,"giffile",,,"GIF Image"
 HKCR,"giffile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"giffile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"giffile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"giffile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"giffile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"giffile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"giffile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,9"
 
 ;; GZIP
@@ -160,13 +161,14 @@
 
 ;; JPE
 HKCR,".jpe",,2,"jpegfile"
 HKCR,".jpe","Content Type",,"image/jpeg"
 HKCR,"jpegfile",,,"JPEG Image"
 HKCR,"jpegfile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"jpegfile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"jpegfile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"jpegfile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"jpegfile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"jpegfile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"jpegfile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,8"
 
 ;; JPEG
@@ -175,24 +177,26 @@
 HKCR,"MIME\Database\Content Type\image/jpeg\Bits","0",1,02,00,00,00,FF,FF,FF,D8
 HKCR,"MIME\Database\Content Type\image/jpeg","Image Filter CLSID",,"%CLSID_IImageDecodeFilter%"
 HKCR,".jpeg",,2,"jpegfile"
 HKCR,".jpeg","Content Type",,"image/jpeg"
 HKCR,"jpegfile",,,"JPEG Image"
 HKCR,"jpegfile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"jpegfile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"jpegfile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"jpegfile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"jpegfile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"jpegfile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"jpegfile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,8"
 
 ;; JPG
 HKCR,".jpg",,2,"jpegfile"
 HKCR,".jpg","Content Type",,"image/jpeg"
 HKCR,"jpegfile",,,"JPEG Image"
 HKCR,"jpegfile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"jpegfile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"jpegfile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"jpegfile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"jpegfile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"jpegfile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"jpegfile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,8"
 
 ;; MHTML
@@ -227,26 +231,28 @@
 HKCR,"MIME\Database\Content Type\image/pjpeg\Bits","0",1,02,00,00,00,FF,FF,FF,D8
 HKCR,"MIME\Database\Content Type\image/pjpeg","Image Filter CLSID",,"%CLSID_IImageDecodeFilter%"
 HKCR,".jfif",,2,"pjpegfile"
 HKCR,".jfif","Content Type",,"image/jpeg"
 HKCR,"pjpegfile",,,"JPEG Image"
 HKCR,"pjpegfile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"pjpegfile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"pjpegfile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"pjpegfile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"pjpegfile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"pjpegfile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"pjpegfile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,8"
 
 ;; PNG
 HKCR,"MIME\Database\Content Type\image/png","Extension",,".png"
 HKCR,"MIME\Database\Content Type\image/x-png","Extension",,".png"
 HKCR,".png",,2,"pngfile"
 HKCR,".png","Content Type",,"image/png"
 HKCR,"pngfile",,,"PNG Image"
 HKCR,"pngfile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"pngfile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"pngfile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"pngfile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"pngfile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"pngfile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"pngfile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,9"
 
 ;; PS
@@ -289,13 +295,14 @@
 ;; XBM
 HKCR,"MIME\Database\Content Type\image/xbm","Extension",,".xbm"
 HKCR,"MIME\Database\Content Type\image/x-xbitmap","CLSID",,"%CLSID_HTMLDocument%"
 HKCR,".xbm","Content Type",,"image/x-xbitmap"
 HKCR,"xbmfile",,,"XBM Image"
 HKCR,"xbmfile\CLSID",,,"%CLSID_HTMLDocument%"
-HKCR,"xbmfile\shell\open\command",,,"""%16422%\Internet Explorer\iexplore.exe"" -nohome"
+; CrossOver hack, bug 7778
+HKCR,"xbmfile\shell\open\command",,,"%11%\winebrowser.exe -nohome"
 HKCR,"xbmfile\shell\open\ddeexec",,,"""file:%%1"",,-1,,,,,"
 HKCR,"xbmfile\shell\open\ddeexec\Application",,,"IExplore"
 HKCR,"xbmfile\shell\open\ddeexec\Topic",,,"WWW_OpenURL"
 ;; HKCR,"xbmfile\DefaultIcon",,,"%16422%\Internet Explorer\iexplore.exe,9"
 
 ;; XHTML
--- a/dlls/mshtml/mshtml_private.h
+++ b/dlls/mshtml/mshtml_private.h
@@ -1155,12 +1155,13 @@
     struct list plugin_hosts;
 };
 
 HRESULT HTMLDocument_Create(IUnknown*,REFIID,void**);
 HRESULT MHTMLDocument_Create(IUnknown*,REFIID,void**);
 HRESULT HTMLLoadOptions_Create(IUnknown*,REFIID,void**);
+HRESULT create_marshaled_doc(HWND,REFIID,void**);
 HRESULT create_document_node(nsIDOMDocument*,GeckoBrowser*,HTMLInnerWindow*,HTMLInnerWindow*,
                              compat_mode_t,HTMLDocumentNode**);
 HRESULT create_doctype_node(HTMLDocumentNode*,nsIDOMNode*,HTMLDOMNode**);
 
 HRESULT create_outer_window(GeckoBrowser*,mozIDOMWindowProxy*,HTMLOuterWindow*,HTMLOuterWindow**);
 HRESULT update_window_doc(HTMLInnerWindow*);
--- a/dlls/mshtml/oleobj.c
+++ b/dlls/mshtml/oleobj.c
@@ -3695,27 +3695,114 @@
     {&IID_IDispatch, &HTMLDocumentObjEvents_data},
     {&IID_IPropertyNotifySink},
     {&DIID_HTMLDocumentEvents, &HTMLDocumentObjEvents_data},
     {&DIID_HTMLDocumentEvents2, &HTMLDocumentObjEvents2_data},
     {NULL}
 };
+
+/* TRUE if we create a dedicated thread for all HTML documents */
+static BOOL gecko_main_thread_config;
+
+static LONG gecko_main_thread;
+static HWND gecko_main_thread_hwnd;
+static HANDLE gecko_main_thread_event;
 
+static DWORD WINAPI gecko_main_thread_proc(void *arg)
+{
+    MSG msg;
+
+    TRACE("\n");
+
+    CoInitialize(NULL);
+
+    gecko_main_thread_hwnd = get_thread_hwnd();
+    if(!gecko_main_thread_hwnd) {
+        ERR("Could not create thread window\n");
+        SetEvent(gecko_main_thread_event);
+        CoUninitialize();
+        return 0;
+    }
+
+    gecko_main_thread = GetCurrentThreadId();
+    SetEvent(gecko_main_thread_event);
+
+    while(GetMessageW(&msg, NULL, 0, 0)) {
+        DispatchMessageW(&msg);
+        TranslateMessage(&msg);
+    }
+
+    CoUninitialize();
+    return 0;
+}
+
+static BOOL WINAPI read_thread_config(INIT_ONCE *once, void *param, void **context)
+{
+    HKEY key;
+    DWORD res;
+    static const WCHAR enable_keyW[] =
+        {'S','o','f','t','w','a','r','e',
+         '\\','W','i','n','e',
+         '\\','M','S','H','T','M','L',
+         '\\','M','a','i','n','T','h','r','e','a','d','H','a','c','k',0};
+
+    res = RegOpenKeyW(HKEY_CURRENT_USER, enable_keyW, &key);
+    if(res == ERROR_SUCCESS) {
+        RegCloseKey(key);
+        FIXME("CXHACK: Using separated main thread.\n");
+        gecko_main_thread_config = TRUE;
+    }
+
+    return TRUE;
+}
+
 static HRESULT create_document_object(BOOL is_mhtml, IUnknown *outer, REFIID riid, void **ppv)
 {
     HTMLDocumentObj *doc;
     HRESULT hres;
 
+    static INIT_ONCE init_once = INIT_ONCE_STATIC_INIT;
+
     if(outer && !IsEqualGUID(&IID_IUnknown, riid)) {
         *ppv = NULL;
         return E_INVALIDARG;
     }
 
     /* ensure that security manager is initialized */
     if(!get_security_manager())
         return E_OUTOFMEMORY;
 
+    /* CXHACK 15579 */
+    InitOnceExecuteOnce(&init_once, read_thread_config, NULL, NULL);
+    if(gecko_main_thread_config && !gecko_main_thread) {
+        HANDLE thread, event;
+
+        event = CreateEventW(NULL, TRUE, FALSE, NULL);
+        if(InterlockedCompareExchangePointer(&gecko_main_thread_event, event, NULL))
+            CloseHandle(event);
+
+        thread = CreateThread(NULL, 0, gecko_main_thread_proc, NULL, 0, NULL);
+        if(thread) {
+            WaitForSingleObject(gecko_main_thread_event, INFINITE);
+            CloseHandle(thread);
+        }else {
+            ERR("Could not create a thread\n");
+        }
+    }
+
+    if(!gecko_main_thread) {
+        gecko_main_thread = GetCurrentThreadId();
+        gecko_main_thread_hwnd = get_thread_hwnd();
+    }else if(GetCurrentThreadId() != gecko_main_thread) {
+        FIXME("CXHACK: Creating HTMLDocument outside Gecko main thread\n");
+        if(!gecko_main_thread_config) {
+            FIXME("CXHACK: Dedicated main thread not configured\n");
+            FIXME("CXHACK: Create HKCU\\Software\\Wine\\MSHTML\\MainThreadHack key\n");
+        }
+        return create_marshaled_doc(gecko_main_thread_hwnd, riid, ppv);
+    }
+
     doc = calloc(1, sizeof(HTMLDocumentObj));
     if(!doc)
         return E_OUTOFMEMORY;
 
     doc->ref = 1;
     doc->IUnknown_inner.lpVtbl = &HTMLDocumentObjVtbl;
--- a/dlls/mshtml/task.c
+++ b/dlls/mshtml/task.c
@@ -30,12 +30,13 @@
 
 #include "mshtml_private.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(mshtml);
 
 #define WM_PROCESSTASK 0x8008
+#define WM_CREATEDOC   0x8018
 #define TIMER_ID 0x3000
 
 typedef struct {
     HTMLInnerWindow *window;
     DWORD id;
     DWORD time;
@@ -374,12 +375,19 @@
 
 done:
     if(!--thread_data->tasks_locked && (!list_empty(&thread_data->task_list) || !list_empty(&thread_data->event_task_list)))
         PostMessageW(thread_data->thread_hwnd, WM_PROCESSTASK, 0, 0);
     return 0;
 }
+
+typedef struct {
+    IUnknown *unk;
+    IID iid;
+    IStream *stream;
+    HRESULT hres;
+} create_doc_params_t;
 
 static LRESULT WINAPI hidden_proc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam)
 {
     thread_data_t *thread_data;
 
     switch(msg) {
@@ -420,13 +428,27 @@
             if(!head)
                 break;
         }
         return 0;
     case WM_TIMER:
         return process_timer();
+    case WM_CREATEDOC: {
+        create_doc_params_t *params = (create_doc_params_t*)lParam;
+        IUnknown *unk;
+
+        TRACE("WM_CREATEDOC %p\n", params);
+
+        params->hres = HTMLDocument_Create(NULL, &params->iid, (void**)&unk);
+        if(FAILED(params->hres))
+            return 0;
+
+        params->hres = CoMarshalInterface(params->stream, &params->iid, unk, MSHCTX_INPROC, NULL, MSHLFLAGS_NORMAL);
+        IUnknown_Release(unk);
+        return 0;
     }
+    }
 
     if(msg > WM_USER)
         FIXME("(%p %d %Ix %Ix)\n", hwnd, msg, wParam, lParam);
 
     return DefWindowProcW(hwnd, msg, wParam, lParam);
 }
@@ -462,12 +484,41 @@
     if(!thread_data->thread_hwnd)
         thread_data->thread_hwnd = create_thread_hwnd();
 
     return thread_data->thread_hwnd;
 }
 
+HRESULT create_marshaled_doc(HWND main_thread_hwnd, REFIID riid, void **ppv)
+{
+    create_doc_params_t params = {NULL, *riid, NULL, E_FAIL};
+    LARGE_INTEGER zero;
+    BOOL res;
+    HRESULT hres;
+
+    hres = CreateStreamOnHGlobal(NULL, TRUE, &params.stream);
+    if(FAILED(hres))
+        return hres;
+
+    res = SendMessageW(main_thread_hwnd, WM_CREATEDOC, 0, (LPARAM)&params);
+    TRACE("SendMessage ret %x\n", res);
+    if(FAILED(params.hres)) {
+        WARN("EM_CREATEDOC failed: %08lx\n", params.hres);
+        IStream_Release(params.stream);
+        return hres;
+    }
+
+    zero.QuadPart = 0;
+    hres = IStream_Seek(params.stream, zero, STREAM_SEEK_SET, NULL);
+    if(SUCCEEDED(hres))
+        hres = CoUnmarshalInterface(params.stream, riid, ppv);
+    IStream_Release(params.stream);
+    if(FAILED(hres))
+        WARN("CoUnmarshalInterface failed: %08lx\n", hres);
+    return hres;
+}
+
 thread_data_t *get_thread_data(BOOL create)
 {
     thread_data_t *thread_data;
 
     if(mshtml_tls == TLS_OUT_OF_INDEXES) {
         DWORD tls;
--- a/dlls/mshtml/tests/style.c
+++ b/dlls/mshtml/tests/style.c
@@ -723,13 +723,13 @@
 
     hres = IHTMLStyle5_put_maxHeight(style5, vdefault);
     ok(hres == S_OK, "put maxHeight failed:%08lx\n", hres);
     VariantClear(&vdefault);
 }
 
-static void test_style6(IHTMLStyle6 *style)
+static void test_style6(IHTMLStyle6 *style, IHTMLCSSStyleDeclaration *css_style)
 {
     BSTR str;
     HRESULT hres;
 
     str = (void*)0xdeadbeef;
     hres = IHTMLStyle6_get_outline(style, &str);
@@ -779,12 +779,54 @@
 
     str = NULL;
     hres = IHTMLStyle6_get_borderSpacing(style, &str);
     ok(hres == S_OK, "get_borderSpacing failed: %08lx\n", hres);
     ok(!lstrcmpW(str, L"10px"), "borderSpacing = %s\n", wine_dbgstr_w(str));
     SysFreeString(str);
+
+    hres = IHTMLStyle6_get_content(style, &str);
+    ok(hres == S_OK, "get_content failed: %08lx\n", hres);
+    ok(!str, "content = %s\n", debugstr_w(str));
+
+    if(css_style) {
+        hres = IHTMLCSSStyleDeclaration_get_content(css_style, &str);
+        ok(hres == S_OK, "get_content failed: %08lx\n", hres);
+        ok(!str, "content = %s\n", debugstr_w(str));
+    }
+
+    str = SysAllocString(L"normal");
+    hres = IHTMLStyle6_put_content(style, str);
+    ok(hres == S_OK, "put_content failed: %08lx\n", hres);
+    SysFreeString(str);
+
+    hres = IHTMLStyle6_get_content(style, &str);
+    ok(hres == S_OK, "get_content failed: %08lx\n", hres);
+    ok(!wcscmp(str, L"normal"), "content = %s\n", debugstr_w(str));
+    SysFreeString(str);
+
+    if(css_style) {
+        hres = IHTMLCSSStyleDeclaration_get_content(css_style, &str);
+        ok(hres == S_OK, "get_content failed: %08lx\n", hres);
+        ok(!wcscmp(str, L"normal"), "content = %s\n", debugstr_w(str));
+        SysFreeString(str);
+
+        str = SysAllocString(L"none");
+        hres = IHTMLCSSStyleDeclaration_put_content(css_style, str);
+        ok(hres == S_OK, "put_content failed: %08lx\n", hres);
+        SysFreeString(str);
+
+        hres = IHTMLCSSStyleDeclaration_get_content(css_style, &str);
+        ok(hres == S_OK, "get_content failed: %08lx\n", hres);
+        ok(!wcscmp(str, L"none"), "content = %s\n", debugstr_w(str));
+        SysFreeString(str);
+
+        hres = IHTMLStyle6_get_content(style, &str);
+        ok(hres == S_OK, "get_content failed: %08lx\n", hres);
+        ok(!wcscmp(str, L"none"), "content = %s\n", debugstr_w(str));
+        SysFreeString(str);
+    }
 }
 
 static void test_css_style_declaration(IHTMLCSSStyleDeclaration *css_style)
 {
     VARIANT v;
     BSTR str;
@@ -3340,13 +3382,13 @@
     }else {
         win_skip("IHTMLStyle5 not available\n");
     }
 
     hres = IHTMLStyle_QueryInterface(style, &IID_IHTMLStyle6, (void**)&style6);
     if(SUCCEEDED(hres)) {
-        test_style6(style6);
+        test_style6(style6, css_style);
         IHTMLStyle6_Release(style6);
     }else {
         win_skip("IHTMLStyle6 not available\n");
     }
 
     if(compat_mode >= COMPAT_IE9) {
