--- a/dlls/combase/apartment.c
+++ b/dlls/combase/apartment.c
@@ -108,12 +108,30 @@
     }
     LeaveCriticalSection(&dlls_cs);
 
     return ret;
 }
 
+static BOOL is_olmapi32( const WCHAR *library_name )
+{
+    static const WCHAR olmapi32[] = {'o','l','m','a','p','i','3','2','.','d','l','l',0};
+    DWORD len = sizeof(olmapi32)/sizeof(olmapi32[0]) - 1, len2 = lstrlenW( library_name );
+    return (len <= len2 && !wcsicmp( library_name + len2 - len, olmapi32 ));
+}
+
+static BOOL is_outlook(void)
+{
+    static const WCHAR ol2010[] =
+        {'o','f','f','i','c','e','1','4','\\','o','u','t','l','o','o','k','.','e','x','e',0};
+    static const WCHAR ol2013[] =
+        {'o','f','f','i','c','e','1','5','\\','o','u','t','l','o','o','k','.','e','x','e',0};
+    WCHAR path[MAX_PATH];
+    DWORD len = sizeof(ol2010)/sizeof(ol2010[0]) - 1, len2 = GetModuleFileNameW( NULL, path, MAX_PATH );
+    return (len <= len2 && (!wcsicmp( path + len2 - len, ol2010 ) || !wcsicmp( path + len2 - len, ol2013 )));
+}
+
 /* caller must ensure that library_name is not already in the open dll list */
 static HRESULT apartment_add_dll(const WCHAR *library_name, struct opendll **ret)
 {
     struct opendll *entry;
     int len;
     HRESULT hr = S_OK;
@@ -130,12 +148,19 @@
     hLibrary = LoadLibraryExW(library_name, 0, LOAD_WITH_ALTERED_SEARCH_PATH);
     if (!hLibrary)
     {
         ERR("couldn't load in-process dll %s\n", debugstr_w(library_name));
         /* failure: DLL could not be loaded */
         return E_ACCESSDENIED; /* FIXME: or should this be CO_E_DLLNOTFOUND? */
+    }
+
+    if (is_olmapi32( library_name ) && is_outlook()) /* CrossOver hack bug 14243 */
+    {
+        HRESULT (WINAPI *pMAPIInitialize)(void *) = (void *)GetProcAddress( hLibrary, "MAPIInitialize" );
+        FIXME( "HACK: calling olmapi32.MAPIInitialize\n" );
+        pMAPIInitialize( NULL );
     }
 
     /* DllCanUnloadNow is optional */
     DllCanUnloadNow = (void *)GetProcAddress(hLibrary, "DllCanUnloadNow");
     DllGetClassObject = (void *)GetProcAddress(hLibrary, "DllGetClassObject");
     if (!DllGetClassObject)
--- a/dlls/combase/combase.c
+++ b/dlls/combase/combase.c
@@ -2741,13 +2741,16 @@
     if (!apt)
     {
         ERR("apartment not initialised\n");
         return;
     }
 
-    apartment_freeunusedlibraries(apt, unload_delay);
+    /* CODEWEAVERS HACK for Office 2003: Don't call
+     * apartment_freeunusedlibraries as it appears to do something the app
+     * doesn't expect and free a module that is in-use */
+    if (0) apartment_freeunusedlibraries(apt, unload_delay);
 }
 
 /*
  * When locked, don't modify list (unless we add a new head), so that it's
  * safe to iterate it. Freeing of list entries is delayed and done on unlock.
  */
